{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "from sqlalchemy import create_engine\n",
    "import pandas as pd\n",
    "import os\n",
    "from os import path\n",
    "import re\n",
    "import pymysql\n",
    "import matplotlib\n",
    "matplotlib.use('Agg')\n",
    "from matplotlib import pyplot as plt\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "from tqdm import tqdm\n",
    "from multiprocessing import Pool\n",
    "import yaml\n",
    "\n",
    "with open('../config.yml') as f:\n",
    "    config = yaml.load(f)\n",
    "\n",
    "project_dir = config['PROJECT']\n",
    "project_db_dir = config['DATABASE']\n",
    "\n",
    "def imap_unordered_bar(func, args, number_of_elements, n_processes = 30):\n",
    "    scores = []\n",
    "    p = Pool(n_processes)\n",
    "\n",
    "    with tqdm(total = number_of_elements) as pbar:\n",
    "        for res in tqdm(p.imap_unordered(func, args, chunksize=n_processes)):\n",
    "            scores += res\n",
    "            pbar.update()\n",
    "\n",
    "    pbar.close()\n",
    "    p.close()\n",
    "    p.join()\n",
    "\n",
    "    return scores\n",
    "\n",
    "engine = create_engine(\"mysql+pymysql://{user}:{pw}@{server}/{db}\"\n",
    "                       .format(server=config['MYSQL_HOST'], user=config['MYSQL_USER'], pw=config['MYSQL_PASSWD'], \n",
    "                               db=config['MYSQL_DB']))\n",
    "con = pymysql.connect(host=config['MYSQL_HOST'], user=config['MYSQL_USER'], passwd=config['MYSQL_PASSWD'], \n",
    "                      db=config['MYSQL_DB'])\n",
    "\n",
    "cur = con.cursor()\n",
    "\n",
    "plot_dir = project_dir+'/tmp/clinvar/'\n",
    "alignment_dir = project_dir+\"/results/alignments/\"\n",
    "\n",
    "polarity = {\n",
    "\t'A': 'nonpolar',\n",
    "    'B': '-',\n",
    "    'Z': '-',\n",
    "\t'R': 'polar',\n",
    "\t'N': 'polar',\n",
    "\t'D': 'polar',\n",
    "\t'C': 'nonpolar',\n",
    "\t'E': 'polar',\n",
    "\t'Q': 'polar',\n",
    "\t'G': 'nonpolar',\n",
    "\t'H': 'polar',\n",
    "\t'I': 'nonpolar',\n",
    "\t'L': 'nonpolar',\n",
    "\t'K': 'polar',\n",
    "\t'M': 'nonpolar',\n",
    "\t'F': 'nonpolar',\n",
    "\t'P': 'nonpolar',\n",
    "\t'S': 'polar',\n",
    "\t'T': 'polar',\n",
    "\t'W': 'nonpolar',\n",
    "\t'Y': 'polar',\n",
    "\t'V': 'nonpolar',\n",
    "\t'U': 'polar',\n",
    "\t'-': '-'\n",
    "}\n",
    "def sqlize(significance_list):\n",
    "\treturn \"'\"+ \"', '\".join(significance_list) + \"'\"\n",
    "\n",
    "ncbi_to_transcript_id = pd.read_sql(\"SELECT gdb.db_id AS NCBI, gdb.convart_gene_id AS id, ncbi.meta_value AS gene_symbol \"\n",
    "\t\t\t\t\t\t\t\t\"FROM convart_gene_to_db AS gdb \" +\n",
    "\t\t\t\t\t\t\t\t\"LEFT JOIN ncbi_gene_meta AS ncbi ON ncbi.ncbi_gene_id = gdb.db_id WHERE gdb.db='NCBI' \"+\n",
    "\t\t\t\t\t\t\t\t\"AND ncbi.meta_key = 'gene_symbol' GROUP BY gdb.convart_gene_id\", con)\n",
    "ncbi_to_transcript_id.set_index('id', drop=False, inplace=True)\n",
    "\n",
    "\n",
    "genes = pd.read_sql(\"SELECT DISTINCT g.id, fasta, sequence, m.id as msa_id \"+\n",
    "\t\t\t\"FROM convart_gene AS g \"+\n",
    "\t\t\t\"INNER JOIN msa_best_combination AS mb ON mb.convart_gene_id = g.id \" +\n",
    "\t\t\t\"INNER JOIN msa AS m ON m.id = mb.msa_id \"\n",
    "\t\t\t\"WHERE g.species_id='Homo sapiens'\", con)\n",
    "\n",
    "\n",
    "benign_list = sqlize([\"Benign\",\"Benign, association\",\"Benign/Likely benign\",\"Benign/Likely benign, Affects\",\"Benign/Likely benign, association\",\n",
    "\t\t\t\"Benign/Likely benign, drug response\",\"Benign/Likely benign, drug response, risk factor\",\"Benign/Likely benign, other\",\"Benign/Likely benign, protective\",\n",
    "\t\t\t\"Benign/Likely benign, protective, risk factor\",\"Benign/Likely benign, risk factor\",\"Benign, other\",\"Benign, risk factor\",\"Likely benign\",\n",
    "\t\t\t\"Likely benign, drug response, other\",\"Likely benign, other\",\"Likely benign, risk factor\"])\n",
    "vus_list = sqlize([\"Uncertain significance\",\"Uncertain significance, drug response\",\"Uncertain significance, other\",\"Uncertain significance, risk factor\"])\n",
    "pathogenic_list = sqlize([\"Pathogenic\",\"Pathogenic, Affects\",\"Pathogenic, association, protective\",\"Pathogenic, drug response\",\"Pathogenic/Likely pathogenic\",\n",
    "\"Pathogenic/Likely pathogenic, drug response\",\"Pathogenic/Likely pathogenic, other\",\"Pathogenic/Likely pathogenic, risk factor\",\"Pathogenic, other\",\n",
    "\"Pathogenic, other, risk factor\",\"Pathogenic, protective\",\"Pathogenic, risk factor\",\"Likely pathogenic\",\"Likely pathogenic, association\",\n",
    "\"Likely pathogenic, drug response\",\"Likely pathogenic, other\",\"Likely pathogenic, risk factor\"])\n",
    "\n",
    "ptm_data = pd.read_sql(\"SELECT cdb.convart_gene_id AS transcript_id, GROUP_CONCAT(DISTINCT position) as positions\"\n",
    "\t\t\t\t\t+\", COUNT(DISTINCT position) as variation_count FROM ptm INNER JOIN convart_gene_to_db AS cdb ON cdb.db_id=ptm.acc_id  WHERE position > 0 \"\n",
    "                    +\"GROUP BY cdb.convart_gene_id\", con)\n",
    "clinvar_data = pd.read_sql(\"SELECT gdb.convart_gene_id AS transcript_id, GROUP_CONCAT(DISTINCT position) as positions\"\n",
    "\t\t\t+\", COUNT(DISTINCT position) as variation_count FROM clinvar INNER JOIN convart_gene_to_db AS gdb ON gdb.db_id=np_id WHERE position > 0 \"\n",
    "                           +\"GROUP BY gdb.convart_gene_id\", con)\n",
    "pathogenic_clinvar_data = pd.read_sql(\"SELECT gdb.convart_gene_id AS transcript_id, GROUP_CONCAT(DISTINCT position) as positions, COUNT(DISTINCT position) \"\n",
    "\t\t\t+\"as variation_count FROM clinvar INNER JOIN convart_gene_to_db AS gdb ON gdb.db_id=np_id WHERE position > 0 AND clinical_significance IN (\"+pathogenic_list+\")\"+\n",
    "\t\t\t\" GROUP BY gdb.convart_gene_id\", con)\n",
    "vus_clinvar_data = pd.read_sql(\"SELECT gdb.convart_gene_id AS transcript_id, GROUP_CONCAT(DISTINCT position) as positions, COUNT(DISTINCT position) \"\n",
    "\t\t\t+\"as variation_count FROM clinvar INNER JOIN convart_gene_to_db AS gdb ON gdb.db_id=np_id WHERE position > 0 AND clinical_significance IN (\"+vus_list+\")\"+\n",
    "\t\t\t\" GROUP BY gdb.convart_gene_id\", con)\n",
    "benign_clinvar_data = pd.read_sql(\"SELECT gdb.convart_gene_id AS transcript_id, GROUP_CONCAT(DISTINCT position) as positions, COUNT(DISTINCT position) \"\n",
    "\t\t\t+\"as variation_count FROM clinvar INNER JOIN convart_gene_to_db AS gdb ON gdb.db_id=np_id WHERE position > 0 AND clinical_significance IN (\"+benign_list+\")\"+\n",
    "\t\t\t\" GROUP BY gdb.convart_gene_id\", con)\n",
    "gnomad_data = pd.read_sql(\"SELECT gdb.convart_gene_id AS transcript_id, GROUP_CONCAT(DISTINCT position) as positions\"\n",
    "\t\t\t\t\t+\", COUNT(DISTINCT position) as variation_count FROM gnomad INNER JOIN convart_gene_to_db AS gdb ON gdb.db_id=canonical_transcript WHERE position > 0 \"+\n",
    "                     \" GROUP BY gdb.convart_gene_id\", con)\n",
    "\n",
    "cosmic_data = pd.read_sql(\"SELECT gdb.convart_gene_id AS transcript_id, GROUP_CONCAT(DISTINCT position) as positions\"\n",
    "\t\t\t\t\t+\", COUNT(DISTINCT position) as variation_count FROM CosmicMutantExport INNER JOIN convart_gene_to_db AS gdb ON gdb.db_id=accession_number WHERE position > 0 \"+\n",
    "                     \" GROUP BY gdb.convart_gene_id\", con)\n",
    "dbsnp_data = pd.read_sql(\"SELECT Feature AS transcript_id, GROUP_CONCAT(DISTINCT Protein_position) as positions\"\n",
    "\t\t\t\t\t+\", COUNT(DISTINCT Protein_position) as variation_count FROM dbsnp INNER JOIN convart_gene_to_db AS gdb ON gdb.db_id=Feature WHERE Protein_position > 0 \"+\n",
    "                     \" GROUP BY gdb.convart_gene_id\", con)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_data = pd.concat([clinvar_data, gnomad_data, cosmic_data, dbsnp_data], axis=0, ignore_index=True)\n",
    "combined_data = combined_data.groupby('transcript_id').apply(lambda x: pd.Series({'positions':','.join([str(i) for i in x['positions']]), \n",
    "                                                                  'variation_count': sum(x['variation_count'])}))\n",
    "combined_data = combined_data.reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|          | 0/55428 [00:00<?, ?it/s]\n",
      "  0%|          | 1/55428 [00:00<9:01:56,  1.70it/s]\n",
      "  0%|          | 211/55428 [00:00<6:18:03,  2.43it/s]\n",
      "  1%|          | 571/55428 [00:00<4:23:00,  3.48it/s]\n",
      "  1%|▏         | 781/55428 [00:00<3:03:32,  4.96it/s]\n",
      "  2%|▏         | 1021/55428 [00:01<2:08:02,  7.08it/s]\n",
      "  2%|▏         | 1231/55428 [00:01<1:29:24, 10.10it/s]\n",
      "  3%|▎         | 1501/55428 [00:01<1:02:23, 14.41it/s]\n",
      "  3%|▎         | 1741/55428 [00:01<43:35, 20.52it/s]  \n",
      "  4%|▎         | 1947/55428 [00:01<30:32, 29.19it/s]\n",
      "  4%|▍         | 2191/55428 [00:01<21:23, 41.48it/s]\n",
      "  4%|▍         | 2406/55428 [00:01<15:02, 58.73it/s]\n",
      "  5%|▍         | 2702/55428 [00:01<10:33, 83.19it/s]\n",
      "  5%|▌         | 2971/55428 [00:01<07:28, 116.88it/s]\n",
      "  6%|▌         | 3199/55428 [00:02<05:24, 161.19it/s]\n",
      "  6%|▌         | 3401/55428 [00:02<03:55, 221.27it/s]\n",
      "  7%|▋         | 3662/55428 [00:02<02:49, 304.87it/s]\n",
      "  7%|▋         | 4081/55428 [00:02<02:02, 420.79it/s]\n",
      "  8%|▊         | 4341/55428 [00:02<01:33, 544.94it/s]\n",
      "  8%|▊         | 4576/55428 [00:02<01:13, 695.38it/s]\n",
      "  9%|▉         | 4921/55428 [00:02<00:55, 907.51it/s]\n",
      "  9%|▉         | 5172/55428 [00:02<00:46, 1092.37it/s]\n",
      " 10%|▉         | 5410/55428 [00:03<00:40, 1221.25it/s]\n",
      " 10%|█         | 5761/55428 [00:03<00:33, 1495.70it/s]\n",
      " 11%|█         | 6008/55428 [00:03<00:29, 1676.98it/s]\n",
      " 11%|█▏        | 6251/55428 [00:03<00:29, 1639.48it/s]\n",
      " 12%|█▏        | 6468/55428 [00:03<00:28, 1699.95it/s]\n",
      " 13%|█▎        | 6992/55428 [00:03<00:26, 1835.97it/s]\n",
      " 13%|█▎        | 7261/55428 [00:03<00:24, 1999.86it/s]\n",
      " 13%|█▎        | 7480/55428 [00:03<00:23, 1998.98it/s]\n",
      " 14%|█▍        | 7693/55428 [00:04<00:26, 1783.31it/s]\n",
      " 14%|█▍        | 7951/55428 [00:04<00:24, 1941.10it/s]\n",
      " 15%|█▍        | 8281/55428 [00:04<00:23, 2002.37it/s]\n",
      " 15%|█▌        | 8492/55428 [00:04<00:23, 1993.86it/s]\n",
      " 16%|█▌        | 8701/55428 [00:04<00:23, 2001.80it/s]\n",
      " 16%|█▌        | 8971/55428 [00:04<00:21, 2155.68it/s]\n",
      " 17%|█▋        | 9194/55428 [00:04<00:22, 2077.55it/s]\n",
      " 17%|█▋        | 9408/55428 [00:04<00:23, 1933.03it/s]\n",
      " 17%|█▋        | 9691/55428 [00:05<00:21, 2119.21it/s]\n",
      " 18%|█▊        | 9913/55428 [00:05<00:21, 2090.86it/s]\n",
      " 18%|█▊        | 10141/55428 [00:05<00:21, 2140.29it/s]\n",
      " 19%|█▉        | 10441/55428 [00:05<00:19, 2323.00it/s]\n",
      " 19%|█▉        | 10682/55428 [00:05<00:20, 2175.01it/s]\n",
      " 20%|█▉        | 11011/55428 [00:05<00:18, 2381.30it/s]\n",
      " 20%|██        | 11261/55428 [00:05<00:20, 2142.19it/s]\n",
      " 21%|██        | 11488/55428 [00:05<00:22, 1975.86it/s]\n",
      " 21%|██▏       | 11821/55428 [00:06<00:19, 2200.12it/s]\n",
      " 22%|██▏       | 12061/55428 [00:06<00:20, 2160.57it/s]\n",
      " 22%|██▏       | 12301/55428 [00:06<00:20, 2113.15it/s]\n",
      " 23%|██▎       | 12521/55428 [00:06<00:20, 2072.11it/s]\n",
      " 23%|██▎       | 12901/55428 [00:06<00:18, 2354.66it/s]\n",
      " 24%|██▎       | 13154/55428 [00:06<00:19, 2122.06it/s]\n",
      " 24%|██▍       | 13383/55428 [00:06<00:19, 2166.74it/s]\n",
      " 25%|██▍       | 13612/55428 [00:06<00:19, 2179.90it/s]\n",
      " 25%|██▍       | 13839/55428 [00:06<00:19, 2119.98it/s]\n",
      " 25%|██▌       | 14131/55428 [00:07<00:18, 2174.40it/s]\n",
      " 26%|██▌       | 14354/55428 [00:07<00:19, 2145.06it/s]\n",
      " 26%|██▋       | 14572/55428 [00:07<00:20, 2006.10it/s]\n",
      " 27%|██▋       | 14821/55428 [00:07<00:19, 2128.84it/s]\n",
      " 27%|██▋       | 15039/55428 [00:07<00:19, 2038.20it/s]\n",
      " 28%|██▊       | 15361/55428 [00:07<00:17, 2256.82it/s]\n",
      " 28%|██▊       | 15597/55428 [00:07<00:18, 2173.52it/s]\n",
      " 29%|██▊       | 15871/55428 [00:07<00:17, 2290.52it/s]\n",
      "15871it [00:07, 2281.30it/s]\u001b[A\n",
      " 29%|██▉       | 16108/55428 [00:07<00:18, 2075.27it/s]\n",
      " 30%|██▉       | 16441/55428 [00:08<00:17, 2287.62it/s]\n",
      " 30%|███       | 16741/55428 [00:08<00:17, 2180.58it/s]\n",
      " 31%|███       | 16970/55428 [00:08<00:18, 2037.88it/s]\n",
      " 31%|███       | 17281/55428 [00:08<00:16, 2251.62it/s]\n",
      " 32%|███▏      | 17521/55428 [00:08<00:16, 2283.29it/s]\n",
      " 32%|███▏      | 17759/55428 [00:08<00:19, 1896.80it/s]\n",
      " 32%|███▏      | 17967/55428 [00:08<00:19, 1883.08it/s]\n",
      " 33%|███▎      | 18241/55428 [00:08<00:17, 2068.14it/s]\n",
      " 34%|███▎      | 18571/55428 [00:09<00:15, 2315.57it/s]\n",
      " 34%|███▍      | 18823/55428 [00:09<00:18, 1980.74it/s]\n",
      " 34%|███▍      | 19111/55428 [00:09<00:17, 2059.24it/s]\n",
      " 35%|███▍      | 19334/55428 [00:09<00:17, 2075.19it/s]\n",
      " 35%|███▌      | 19553/55428 [00:09<00:18, 1970.02it/s]\n",
      " 36%|███▌      | 19831/55428 [00:09<00:17, 2058.48it/s]\n",
      " 36%|███▋      | 20131/55428 [00:09<00:15, 2268.61it/s]\n",
      " 37%|███▋      | 20370/55428 [00:09<00:17, 1974.75it/s]\n",
      " 37%|███▋      | 20641/55428 [00:10<00:17, 2045.33it/s]\n",
      " 38%|███▊      | 20971/55428 [00:10<00:15, 2199.87it/s]\n",
      " 38%|███▊      | 21211/55428 [00:10<00:18, 1836.33it/s]\n",
      " 39%|███▊      | 21451/55428 [00:10<00:17, 1955.88it/s]\n",
      " 39%|███▉      | 21691/55428 [00:10<00:16, 2055.59it/s]\n",
      " 40%|███▉      | 21961/55428 [00:10<00:15, 2119.84it/s]\n",
      " 40%|████      | 22182/55428 [00:10<00:18, 1809.98it/s]\n",
      " 40%|████      | 22441/55428 [00:10<00:16, 1980.48it/s]\n",
      " 41%|████      | 22741/55428 [00:11<00:14, 2193.57it/s]\n",
      " 41%|████▏     | 22978/55428 [00:11<00:16, 2005.26it/s]\n",
      " 42%|████▏     | 23194/55428 [00:11<00:18, 1787.67it/s]\n",
      " 43%|████▎     | 23581/55428 [00:11<00:15, 2118.26it/s]\n",
      " 43%|████▎     | 23829/55428 [00:11<00:17, 1857.31it/s]\n",
      " 43%|████▎     | 24046/55428 [00:11<00:17, 1752.70it/s]\n",
      " 44%|████▍     | 24301/55428 [00:11<00:16, 1890.15it/s]\n",
      " 44%|████▍     | 24571/55428 [00:12<00:15, 2054.14it/s]\n",
      " 45%|████▍     | 24794/55428 [00:12<00:15, 1980.11it/s]\n",
      " 45%|████▌     | 25081/55428 [00:12<00:13, 2179.63it/s]\n",
      " 46%|████▌     | 25315/55428 [00:12<00:14, 2119.35it/s]\n",
      " 46%|████▌     | 25538/55428 [00:12<00:14, 2091.42it/s]\n",
      " 46%|████▋     | 25755/55428 [00:12<00:14, 2061.79it/s]\n",
      " 47%|████▋     | 26131/55428 [00:12<00:12, 2349.29it/s]\n",
      " 48%|████▊     | 26384/55428 [00:12<00:14, 1993.06it/s]\n",
      " 48%|████▊     | 26606/55428 [00:12<00:14, 1948.27it/s]\n",
      " 48%|████▊     | 26817/55428 [00:13<00:14, 1933.85it/s]\n",
      " 49%|████▉     | 27121/55428 [00:13<00:13, 2142.38it/s]\n",
      " 49%|████▉     | 27361/55428 [00:13<00:13, 2113.08it/s]\n",
      " 50%|████▉     | 27583/55428 [00:13<00:13, 2015.66it/s]\n",
      " 50%|█████     | 27901/55428 [00:13<00:12, 2205.45it/s]\n",
      " 51%|█████     | 28133/55428 [00:13<00:12, 2217.53it/s]\n",
      " 51%|█████     | 28363/55428 [00:13<00:12, 2104.06it/s]\n",
      " 52%|█████▏    | 28581/55428 [00:13<00:12, 2094.99it/s]\n",
      " 52%|█████▏    | 28796/55428 [00:13<00:13, 2021.85it/s]\n",
      " 52%|█████▏    | 29002/55428 [00:14<00:13, 1942.77it/s]\n",
      " 53%|█████▎    | 29341/55428 [00:14<00:12, 2160.25it/s]\n",
      " 53%|█████▎    | 29581/55428 [00:14<00:12, 2118.36it/s]\n",
      " 54%|█████▍    | 29851/55428 [00:14<00:11, 2260.90it/s]\n",
      " 54%|█████▍    | 30091/55428 [00:14<00:11, 2263.90it/s]\n",
      " 55%|█████▍    | 30331/55428 [00:14<00:11, 2188.29it/s]\n",
      " 55%|█████▌    | 30555/55428 [00:14<00:11, 2102.78it/s]\n",
      " 56%|█████▌    | 30770/55428 [00:14<00:11, 2084.52it/s]\n",
      " 56%|█████▌    | 31021/55428 [00:14<00:11, 2145.41it/s]\n",
      " 56%|█████▋    | 31238/55428 [00:15<00:11, 2097.36it/s]\n",
      " 57%|█████▋    | 31531/55428 [00:15<00:11, 2080.12it/s]\n",
      " 57%|█████▋    | 31801/55428 [00:15<00:11, 2106.29it/s]\n",
      " 58%|█████▊    | 32071/55428 [00:15<00:10, 2216.05it/s]\n",
      " 58%|█████▊    | 32341/55428 [00:15<00:09, 2310.42it/s]\n",
      " 59%|█████▉    | 32575/55428 [00:15<00:09, 2311.03it/s]\n",
      " 59%|█████▉    | 32808/55428 [00:15<00:10, 2141.88it/s]\n",
      " 60%|█████▉    | 33031/55428 [00:15<00:10, 2163.57it/s]\n",
      " 60%|██████    | 33331/55428 [00:16<00:09, 2283.14it/s]\n",
      " 61%|██████    | 33563/55428 [00:16<00:11, 1971.88it/s]\n",
      " 61%|██████    | 33771/55428 [00:16<00:11, 1925.83it/s]\n",
      " 61%|██████▏   | 34083/55428 [00:16<00:09, 2175.64it/s]\n",
      " 62%|██████▏   | 34317/55428 [00:16<00:10, 2077.19it/s]\n",
      " 62%|██████▏   | 34591/55428 [00:16<00:09, 2181.92it/s]\n",
      " 63%|██████▎   | 34831/55428 [00:16<00:09, 2206.08it/s]\n",
      " 63%|██████▎   | 35101/55428 [00:16<00:09, 2187.43it/s]\n",
      " 64%|██████▍   | 35371/55428 [00:16<00:08, 2298.30it/s]\n",
      " 64%|██████▍   | 35606/55428 [00:17<00:09, 2178.42it/s]\n",
      " 65%|██████▍   | 35829/55428 [00:17<00:09, 2088.78it/s]\n",
      " 65%|██████▌   | 36042/55428 [00:17<00:09, 2092.62it/s]\n",
      " 65%|██████▌   | 36271/55428 [00:17<00:09, 2116.99it/s]\n",
      " 66%|██████▌   | 36485/55428 [00:17<00:08, 2114.82it/s]\n",
      " 66%|██████▌   | 36698/55428 [00:17<00:09, 1960.08it/s]\n",
      " 67%|██████▋   | 36961/55428 [00:17<00:09, 2030.79it/s]\n",
      " 67%|██████▋   | 37201/55428 [00:17<00:08, 2050.02it/s]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 68%|██████▊   | 37531/55428 [00:17<00:07, 2261.56it/s]\n",
      " 68%|██████▊   | 37766/55428 [00:18<00:07, 2210.10it/s]\n",
      " 69%|██████▊   | 37993/55428 [00:18<00:08, 2028.04it/s]\n",
      " 69%|██████▉   | 38203/55428 [00:18<00:08, 2041.46it/s]\n",
      " 69%|██████▉   | 38461/55428 [00:18<00:07, 2172.50it/s]\n",
      " 70%|██████▉   | 38791/55428 [00:18<00:06, 2393.15it/s]\n",
      "38791it [00:18, 2323.54it/s]\u001b[A\n",
      " 70%|███████   | 39042/55428 [00:18<00:07, 2058.54it/s]\n",
      " 71%|███████   | 39264/55428 [00:18<00:08, 1901.92it/s]\n",
      " 71%|███████▏  | 39512/55428 [00:18<00:07, 2044.55it/s]\n",
      " 72%|███████▏  | 39751/55428 [00:19<00:07, 2119.38it/s]\n",
      " 72%|███████▏  | 39973/55428 [00:19<00:07, 2143.30it/s]\n",
      " 73%|███████▎  | 40194/55428 [00:19<00:07, 2103.60it/s]\n",
      " 73%|███████▎  | 40412/55428 [00:19<00:07, 2111.50it/s]\n",
      " 73%|███████▎  | 40627/55428 [00:19<00:07, 1859.09it/s]\n",
      " 74%|███████▍  | 40981/55428 [00:19<00:07, 2024.55it/s]\n",
      " 74%|███████▍  | 41251/55428 [00:19<00:06, 2183.16it/s]\n",
      " 75%|███████▍  | 41480/55428 [00:19<00:07, 1940.57it/s]\n",
      " 76%|███████▌  | 41851/55428 [00:19<00:06, 2236.21it/s]\n",
      " 76%|███████▌  | 42100/55428 [00:20<00:06, 1971.34it/s]\n",
      " 76%|███████▋  | 42321/55428 [00:20<00:06, 2013.75it/s]\n",
      " 77%|███████▋  | 42901/55428 [00:20<00:05, 2227.97it/s]\n",
      " 78%|███████▊  | 43141/55428 [00:20<00:05, 2195.76it/s]\n",
      " 78%|███████▊  | 43372/55428 [00:20<00:05, 2200.66it/s]\n",
      "43372it [00:20, 2158.48it/s]\u001b[A\n",
      " 79%|███████▊  | 43600/55428 [00:20<00:06, 1887.95it/s]\n",
      " 79%|███████▉  | 43981/55428 [00:20<00:05, 2196.09it/s]\n",
      " 80%|███████▉  | 44229/55428 [00:21<00:05, 2151.12it/s]\n",
      " 80%|████████  | 44464/55428 [00:21<00:05, 1860.15it/s]\n",
      " 81%|████████  | 44941/55428 [00:21<00:04, 2249.83it/s]\n",
      " 82%|████████▏ | 45218/55428 [00:21<00:04, 2054.22it/s]\n",
      " 82%|████████▏ | 45464/55428 [00:21<00:04, 1995.01it/s]\n",
      " 82%|████████▏ | 45722/55428 [00:21<00:04, 2059.76it/s]\n",
      " 83%|████████▎ | 46051/55428 [00:21<00:04, 2290.24it/s]\n",
      " 84%|████████▎ | 46303/55428 [00:22<00:04, 1948.02it/s]\n",
      " 84%|████████▍ | 46681/55428 [00:22<00:03, 2270.25it/s]\n",
      " 85%|████████▍ | 47041/55428 [00:22<00:04, 1999.92it/s]\n",
      " 86%|████████▌ | 47401/55428 [00:22<00:03, 2280.35it/s]\n",
      " 86%|████████▌ | 47671/55428 [00:22<00:03, 2380.77it/s]\n",
      " 86%|████████▋ | 47936/55428 [00:22<00:03, 2306.63it/s]\n",
      " 87%|████████▋ | 48186/55428 [00:22<00:03, 2034.26it/s]\n",
      " 87%|████████▋ | 48481/55428 [00:22<00:03, 2227.68it/s]\n",
      " 88%|████████▊ | 48751/55428 [00:23<00:02, 2330.45it/s]\n",
      " 88%|████████▊ | 48999/55428 [00:23<00:03, 1963.32it/s]\n",
      " 89%|████████▉ | 49216/55428 [00:23<00:03, 1877.32it/s]\n",
      " 90%|████████▉ | 49711/55428 [00:23<00:02, 2132.61it/s]\n",
      " 90%|█████████ | 49981/55428 [00:23<00:02, 2180.59it/s]\n",
      " 91%|█████████ | 50208/55428 [00:23<00:02, 2132.87it/s]\n",
      " 91%|█████████ | 50461/55428 [00:23<00:02, 2202.39it/s]\n",
      " 92%|█████████▏| 50761/55428 [00:24<00:01, 2363.89it/s]\n",
      " 92%|█████████▏| 51005/55428 [00:24<00:02, 2137.29it/s]\n",
      " 92%|█████████▏| 51228/55428 [00:24<00:01, 2155.36it/s]\n",
      " 93%|█████████▎| 51481/55428 [00:24<00:01, 2188.80it/s]\n",
      " 94%|█████████▎| 51871/55428 [00:24<00:01, 2475.19it/s]\n",
      " 94%|█████████▍| 52135/55428 [00:24<00:02, 1280.36it/s]\n",
      " 94%|█████████▍| 52338/55428 [00:25<00:02, 1417.11it/s]\n",
      " 96%|█████████▌| 52951/55428 [00:25<00:01, 1700.37it/s]\n",
      "52951it [00:24, 1628.86it/s]\u001b[A\n",
      " 96%|█████████▌| 53186/55428 [00:25<00:01, 1426.06it/s]\n",
      " 97%|█████████▋| 53851/55428 [00:25<00:00, 1908.91it/s]\n",
      "53851it [00:25, 1835.71it/s]\u001b[A\n",
      " 98%|█████████▊| 54094/55428 [00:25<00:00, 1913.46it/s]\n",
      " 98%|█████████▊| 54392/55428 [00:25<00:00, 2129.71it/s]\n",
      " 99%|█████████▊| 54638/55428 [00:26<00:00, 1768.28it/s]\n",
      " 99%|█████████▉| 55069/55428 [00:26<00:00, 2012.49it/s]\n",
      "100%|█████████▉| 55306/55428 [00:26<00:00, 1203.89it/s]\n",
      "100%|██████████| 55428/55428 [00:26<00:00, 2064.47it/s]\n"
     ]
    }
   ],
   "source": [
    "\n",
    "variations = {\n",
    "    'ptm': ptm_data,\n",
    "\t'variant': combined_data,\n",
    "\t'pathogenic': pathogenic_clinvar_data,\n",
    "\t'vus': vus_clinvar_data,\n",
    "\t'benign': benign_clinvar_data\n",
    "}\n",
    "\n",
    "for key, variation in variations.items():\n",
    "\tvariation.loc[:, 'transcript_id'] = variation.loc[:, 'transcript_id'].astype(str) \n",
    "\tvariations[key] = variation.set_index('transcript_id', drop=False)\n",
    "\n",
    "def conservation_scores_by_gene(row):\n",
    "\tscores = []\n",
    "\n",
    "\tind, msa_details = row\n",
    "   \n",
    "\tmsa_details = msa_details.to_dict()\n",
    "\tmsa_details['id'] = str(msa_details['id'])\n",
    "    \n",
    "\tsequences = ('\\n' + msa_details['fasta']).split('\\n>')\n",
    "\tsequences = [{\n",
    "\t\t\t'transcript_id': msa_details['id'],\n",
    "\t\t\t'species_id': re.sub(r'^(.*?)\\[([a-zA-Z0-9 ]+?)\\]$', r'\\2', seq.split('\\n')[0]),\n",
    "\t\t\t'fasta': re.sub('^(.*?)\\n', '', seq).replace('\\n','').replace(' ', '')} for seq in sequences]\n",
    "\tsequences = list(filter(lambda x : x['fasta'].replace('\\n','') !='', sequences))\n",
    "\n",
    "\thuman_seq = sequences[0]['fasta']\n",
    "\tseq_length = len(human_seq)\n",
    "\tn_sequences = len(sequences)\n",
    "\t\n",
    "\tvariation_positions_by_data = {}\n",
    "\n",
    "\tfor key, data in variations.items():\n",
    "\t\tvariation_positions_by_data[key] = set()\n",
    "        \n",
    "\t\tif str(msa_details['id']) in data.index:\n",
    "\t\t\tcur_positions = data.loc[msa_details['id'], 'positions'].split(',')\n",
    "\t\t\tcur_positions = set([cur_position for cur_position in cur_positions ])\n",
    "\t\t\t[variation_positions_by_data[key].add(int(cur_position))  for cur_position in cur_positions if cur_position != '']\n",
    "        \n",
    "    \n",
    "\tfor pairwise_id in range(1, len(sequences)):\n",
    "\t\texact_match_score = 0\n",
    "\t\tpolarity_match_score = 0\n",
    "\t\t\n",
    "\t\texact_scores_by_data = {}\n",
    "\t\tpolarity_scores_by_data = {}\n",
    "\n",
    "\t\tfor key, data in variations.items():\n",
    "\t\t\texact_scores_by_data[key] = 0\n",
    "\t\t\tpolarity_scores_by_data[key] = 0\n",
    "\n",
    "\t\tpair = sequences[pairwise_id]\n",
    "\t\tpair_seq = pair['fasta']\n",
    "\t\thuman_seq_position = 0\n",
    "\t\t\n",
    "\t\tfor i in range(seq_length):\n",
    "\t\t\tif len(human_seq) != len(pair_seq):\n",
    "\t\t\t\tprint(len(human_seq), len(pair_seq), msa_details['msa_id'])\n",
    "\t\t\t\tsys.exit(-1)\n",
    "\t\t\tif human_seq[i] == '-':\n",
    "\t\t\t\tcontinue\n",
    "\t\t\thuman_seq_position += 1\n",
    "\n",
    "\t\t\tif human_seq[i] == 'X' or pair_seq[i] == 'X':\n",
    "\t\t\t\tcontinue\n",
    "\n",
    "\t\t\taligned_aminoacids = 1 if len({seq[i] for seq in [human_seq, pair_seq]})==1 else 0\n",
    "\t\t\texact_match_score += aligned_aminoacids\n",
    "\n",
    "\t\t\taligned_aminoacids_polarity = 1 if  len({polarity[seq[i]] for seq in [human_seq, pair_seq]}) == 1 else 0\n",
    "\t\t\tpolarity_match_score += aligned_aminoacids_polarity\n",
    "\n",
    "\t\t\tfor key, positions in variation_positions_by_data.items():\n",
    "\t\t\t\tif human_seq_position not in positions:\n",
    "\t\t\t\t\tcontinue\n",
    "\t\t\t\texact_scores_by_data[key] += aligned_aminoacids\n",
    "\t\t\t\tpolarity_scores_by_data[key] += aligned_aminoacids_polarity\n",
    "\n",
    "\t\tscores.append({'transcript_id': msa_details['id'], 'pairwise_transcript_id': pair['transcript_id'],\n",
    "\t\t\t\t\t\t'pairwise_species_id': pair['species_id'],\n",
    "\t\t\t\t\t\t'score_type': 'exact_match_score', 'score': exact_match_score, \n",
    "\t\t\t\t\t\t'number_of_aminoacids': len(msa_details['sequence']),\n",
    "                      'msa_id': msa_details['msa_id']})\n",
    "\t\t#scores.append({'transcript_id': msa_details['id'], 'pairwise_transcript_id': pair['transcript_id'],\n",
    "\t\t#\t\t\t\t'pairwise_species_id': pair['species_id'],\n",
    "\t\t#\t\t\t\t'score_type': 'polarity_match_score', 'score': polarity_match_score,\n",
    "\t\t#\t\t\t\t'number_of_aminoacids': len(msa_details['sequence'])})\n",
    "\n",
    "\t\t#Variant scores\n",
    "\t\tfor key, exact_score in exact_scores_by_data.items():\n",
    "\t\t\tscores.append({'transcript_id': msa_details['id'], 'pairwise_transcript_id': pair['transcript_id'],\n",
    "\t\t\t\t\t\t'pairwise_species_id': pair['species_id'],\n",
    "\t\t\t\t\t\t'score_type': key+'_exact_match_score', 'score': exact_score, \n",
    "\t\t\t\t\t\t'number_of_aminoacids': len(variation_positions_by_data[key]),\n",
    "                      'msa_id': msa_details['msa_id']})\n",
    "\t\t#for key, polarity_score in polarity_scores_by_data.items():\n",
    "\t\t#\tscores.append({'transcript_id': msa_details['id'], 'pairwise_transcript_id': pair['transcript_id'],\n",
    "\t\t#\t\t\t\t'pairwise_species_id': pair['species_id'],\n",
    "\t\t#\t\t\t\t'score_type': key+'_polarity_match_score', 'score': polarity_score, \n",
    "\t\t#\t\t\t\t'number_of_aminoacids': len(variation_positions_by_data[key])})\n",
    "\n",
    "\treturn scores\n",
    "\n",
    "#conservation_scores_by_gene(list(genes.iterrows())[0])\n",
    "scores = imap_unordered_bar(conservation_scores_by_gene, genes.iterrows(), genes.shape[0], 30)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>msa_id</th>\n",
       "      <th>number_of_aminoacids</th>\n",
       "      <th>pairwise_species_id</th>\n",
       "      <th>pairwise_transcript_id</th>\n",
       "      <th>score</th>\n",
       "      <th>score_type</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>31</td>\n",
       "      <td>309</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>371</td>\n",
       "      <td>269</td>\n",
       "      <td>exact_match_score</td>\n",
       "      <td>371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>31</td>\n",
       "      <td>0</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>371</td>\n",
       "      <td>0</td>\n",
       "      <td>ptm_exact_match_score</td>\n",
       "      <td>371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>31</td>\n",
       "      <td>43</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>371</td>\n",
       "      <td>38</td>\n",
       "      <td>variant_exact_match_score</td>\n",
       "      <td>371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>31</td>\n",
       "      <td>0</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>371</td>\n",
       "      <td>0</td>\n",
       "      <td>pathogenic_exact_match_score</td>\n",
       "      <td>371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>31</td>\n",
       "      <td>0</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>371</td>\n",
       "      <td>0</td>\n",
       "      <td>vus_exact_match_score</td>\n",
       "      <td>371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>31</td>\n",
       "      <td>0</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>371</td>\n",
       "      <td>0</td>\n",
       "      <td>benign_exact_match_score</td>\n",
       "      <td>371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>31</td>\n",
       "      <td>309</td>\n",
       "      <td>Rattus norvegicus</td>\n",
       "      <td>371</td>\n",
       "      <td>269</td>\n",
       "      <td>exact_match_score</td>\n",
       "      <td>371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>31</td>\n",
       "      <td>0</td>\n",
       "      <td>Rattus norvegicus</td>\n",
       "      <td>371</td>\n",
       "      <td>0</td>\n",
       "      <td>ptm_exact_match_score</td>\n",
       "      <td>371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>31</td>\n",
       "      <td>43</td>\n",
       "      <td>Rattus norvegicus</td>\n",
       "      <td>371</td>\n",
       "      <td>38</td>\n",
       "      <td>variant_exact_match_score</td>\n",
       "      <td>371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>31</td>\n",
       "      <td>0</td>\n",
       "      <td>Rattus norvegicus</td>\n",
       "      <td>371</td>\n",
       "      <td>0</td>\n",
       "      <td>pathogenic_exact_match_score</td>\n",
       "      <td>371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>31</td>\n",
       "      <td>0</td>\n",
       "      <td>Rattus norvegicus</td>\n",
       "      <td>371</td>\n",
       "      <td>0</td>\n",
       "      <td>vus_exact_match_score</td>\n",
       "      <td>371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>31</td>\n",
       "      <td>0</td>\n",
       "      <td>Rattus norvegicus</td>\n",
       "      <td>371</td>\n",
       "      <td>0</td>\n",
       "      <td>benign_exact_match_score</td>\n",
       "      <td>371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>31</td>\n",
       "      <td>309</td>\n",
       "      <td>Pan troglodytes</td>\n",
       "      <td>371</td>\n",
       "      <td>296</td>\n",
       "      <td>exact_match_score</td>\n",
       "      <td>371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>31</td>\n",
       "      <td>0</td>\n",
       "      <td>Pan troglodytes</td>\n",
       "      <td>371</td>\n",
       "      <td>0</td>\n",
       "      <td>ptm_exact_match_score</td>\n",
       "      <td>371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>31</td>\n",
       "      <td>43</td>\n",
       "      <td>Pan troglodytes</td>\n",
       "      <td>371</td>\n",
       "      <td>42</td>\n",
       "      <td>variant_exact_match_score</td>\n",
       "      <td>371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>31</td>\n",
       "      <td>0</td>\n",
       "      <td>Pan troglodytes</td>\n",
       "      <td>371</td>\n",
       "      <td>0</td>\n",
       "      <td>pathogenic_exact_match_score</td>\n",
       "      <td>371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>31</td>\n",
       "      <td>0</td>\n",
       "      <td>Pan troglodytes</td>\n",
       "      <td>371</td>\n",
       "      <td>0</td>\n",
       "      <td>vus_exact_match_score</td>\n",
       "      <td>371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>31</td>\n",
       "      <td>0</td>\n",
       "      <td>Pan troglodytes</td>\n",
       "      <td>371</td>\n",
       "      <td>0</td>\n",
       "      <td>benign_exact_match_score</td>\n",
       "      <td>371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>31</td>\n",
       "      <td>309</td>\n",
       "      <td>Macaca mulatta</td>\n",
       "      <td>371</td>\n",
       "      <td>291</td>\n",
       "      <td>exact_match_score</td>\n",
       "      <td>371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>31</td>\n",
       "      <td>0</td>\n",
       "      <td>Macaca mulatta</td>\n",
       "      <td>371</td>\n",
       "      <td>0</td>\n",
       "      <td>ptm_exact_match_score</td>\n",
       "      <td>371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>31</td>\n",
       "      <td>43</td>\n",
       "      <td>Macaca mulatta</td>\n",
       "      <td>371</td>\n",
       "      <td>41</td>\n",
       "      <td>variant_exact_match_score</td>\n",
       "      <td>371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>31</td>\n",
       "      <td>0</td>\n",
       "      <td>Macaca mulatta</td>\n",
       "      <td>371</td>\n",
       "      <td>0</td>\n",
       "      <td>pathogenic_exact_match_score</td>\n",
       "      <td>371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>31</td>\n",
       "      <td>0</td>\n",
       "      <td>Macaca mulatta</td>\n",
       "      <td>371</td>\n",
       "      <td>0</td>\n",
       "      <td>vus_exact_match_score</td>\n",
       "      <td>371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>31</td>\n",
       "      <td>0</td>\n",
       "      <td>Macaca mulatta</td>\n",
       "      <td>371</td>\n",
       "      <td>0</td>\n",
       "      <td>benign_exact_match_score</td>\n",
       "      <td>371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>31</td>\n",
       "      <td>309</td>\n",
       "      <td>Xenopus tropicalis</td>\n",
       "      <td>371</td>\n",
       "      <td>228</td>\n",
       "      <td>exact_match_score</td>\n",
       "      <td>371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>31</td>\n",
       "      <td>0</td>\n",
       "      <td>Xenopus tropicalis</td>\n",
       "      <td>371</td>\n",
       "      <td>0</td>\n",
       "      <td>ptm_exact_match_score</td>\n",
       "      <td>371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>31</td>\n",
       "      <td>43</td>\n",
       "      <td>Xenopus tropicalis</td>\n",
       "      <td>371</td>\n",
       "      <td>33</td>\n",
       "      <td>variant_exact_match_score</td>\n",
       "      <td>371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>31</td>\n",
       "      <td>0</td>\n",
       "      <td>Xenopus tropicalis</td>\n",
       "      <td>371</td>\n",
       "      <td>0</td>\n",
       "      <td>pathogenic_exact_match_score</td>\n",
       "      <td>371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>31</td>\n",
       "      <td>0</td>\n",
       "      <td>Xenopus tropicalis</td>\n",
       "      <td>371</td>\n",
       "      <td>0</td>\n",
       "      <td>vus_exact_match_score</td>\n",
       "      <td>371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>31</td>\n",
       "      <td>0</td>\n",
       "      <td>Xenopus tropicalis</td>\n",
       "      <td>371</td>\n",
       "      <td>0</td>\n",
       "      <td>benign_exact_match_score</td>\n",
       "      <td>371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2062302</th>\n",
       "      <td>55350</td>\n",
       "      <td>552</td>\n",
       "      <td>Pan troglodytes</td>\n",
       "      <td>255344</td>\n",
       "      <td>545</td>\n",
       "      <td>exact_match_score</td>\n",
       "      <td>255344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2062303</th>\n",
       "      <td>55350</td>\n",
       "      <td>13</td>\n",
       "      <td>Pan troglodytes</td>\n",
       "      <td>255344</td>\n",
       "      <td>13</td>\n",
       "      <td>ptm_exact_match_score</td>\n",
       "      <td>255344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2062304</th>\n",
       "      <td>55350</td>\n",
       "      <td>335</td>\n",
       "      <td>Pan troglodytes</td>\n",
       "      <td>255344</td>\n",
       "      <td>329</td>\n",
       "      <td>variant_exact_match_score</td>\n",
       "      <td>255344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2062305</th>\n",
       "      <td>55350</td>\n",
       "      <td>0</td>\n",
       "      <td>Pan troglodytes</td>\n",
       "      <td>255344</td>\n",
       "      <td>0</td>\n",
       "      <td>pathogenic_exact_match_score</td>\n",
       "      <td>255344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2062306</th>\n",
       "      <td>55350</td>\n",
       "      <td>0</td>\n",
       "      <td>Pan troglodytes</td>\n",
       "      <td>255344</td>\n",
       "      <td>0</td>\n",
       "      <td>vus_exact_match_score</td>\n",
       "      <td>255344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2062307</th>\n",
       "      <td>55350</td>\n",
       "      <td>0</td>\n",
       "      <td>Pan troglodytes</td>\n",
       "      <td>255344</td>\n",
       "      <td>0</td>\n",
       "      <td>benign_exact_match_score</td>\n",
       "      <td>255344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2062308</th>\n",
       "      <td>55350</td>\n",
       "      <td>552</td>\n",
       "      <td>Macaca mulatta</td>\n",
       "      <td>255344</td>\n",
       "      <td>522</td>\n",
       "      <td>exact_match_score</td>\n",
       "      <td>255344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2062309</th>\n",
       "      <td>55350</td>\n",
       "      <td>13</td>\n",
       "      <td>Macaca mulatta</td>\n",
       "      <td>255344</td>\n",
       "      <td>11</td>\n",
       "      <td>ptm_exact_match_score</td>\n",
       "      <td>255344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2062310</th>\n",
       "      <td>55350</td>\n",
       "      <td>335</td>\n",
       "      <td>Macaca mulatta</td>\n",
       "      <td>255344</td>\n",
       "      <td>313</td>\n",
       "      <td>variant_exact_match_score</td>\n",
       "      <td>255344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2062311</th>\n",
       "      <td>55350</td>\n",
       "      <td>0</td>\n",
       "      <td>Macaca mulatta</td>\n",
       "      <td>255344</td>\n",
       "      <td>0</td>\n",
       "      <td>pathogenic_exact_match_score</td>\n",
       "      <td>255344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2062312</th>\n",
       "      <td>55350</td>\n",
       "      <td>0</td>\n",
       "      <td>Macaca mulatta</td>\n",
       "      <td>255344</td>\n",
       "      <td>0</td>\n",
       "      <td>vus_exact_match_score</td>\n",
       "      <td>255344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2062313</th>\n",
       "      <td>55350</td>\n",
       "      <td>0</td>\n",
       "      <td>Macaca mulatta</td>\n",
       "      <td>255344</td>\n",
       "      <td>0</td>\n",
       "      <td>benign_exact_match_score</td>\n",
       "      <td>255344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2062314</th>\n",
       "      <td>55350</td>\n",
       "      <td>552</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>255344</td>\n",
       "      <td>403</td>\n",
       "      <td>exact_match_score</td>\n",
       "      <td>255344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2062315</th>\n",
       "      <td>55350</td>\n",
       "      <td>13</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>255344</td>\n",
       "      <td>9</td>\n",
       "      <td>ptm_exact_match_score</td>\n",
       "      <td>255344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2062316</th>\n",
       "      <td>55350</td>\n",
       "      <td>335</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>255344</td>\n",
       "      <td>233</td>\n",
       "      <td>variant_exact_match_score</td>\n",
       "      <td>255344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2062317</th>\n",
       "      <td>55350</td>\n",
       "      <td>0</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>255344</td>\n",
       "      <td>0</td>\n",
       "      <td>pathogenic_exact_match_score</td>\n",
       "      <td>255344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2062318</th>\n",
       "      <td>55350</td>\n",
       "      <td>0</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>255344</td>\n",
       "      <td>0</td>\n",
       "      <td>vus_exact_match_score</td>\n",
       "      <td>255344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2062319</th>\n",
       "      <td>55350</td>\n",
       "      <td>0</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>255344</td>\n",
       "      <td>0</td>\n",
       "      <td>benign_exact_match_score</td>\n",
       "      <td>255344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2062320</th>\n",
       "      <td>55350</td>\n",
       "      <td>552</td>\n",
       "      <td>Xenopus tropicalis</td>\n",
       "      <td>255344</td>\n",
       "      <td>230</td>\n",
       "      <td>exact_match_score</td>\n",
       "      <td>255344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2062321</th>\n",
       "      <td>55350</td>\n",
       "      <td>13</td>\n",
       "      <td>Xenopus tropicalis</td>\n",
       "      <td>255344</td>\n",
       "      <td>7</td>\n",
       "      <td>ptm_exact_match_score</td>\n",
       "      <td>255344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2062322</th>\n",
       "      <td>55350</td>\n",
       "      <td>335</td>\n",
       "      <td>Xenopus tropicalis</td>\n",
       "      <td>255344</td>\n",
       "      <td>135</td>\n",
       "      <td>variant_exact_match_score</td>\n",
       "      <td>255344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2062323</th>\n",
       "      <td>55350</td>\n",
       "      <td>0</td>\n",
       "      <td>Xenopus tropicalis</td>\n",
       "      <td>255344</td>\n",
       "      <td>0</td>\n",
       "      <td>pathogenic_exact_match_score</td>\n",
       "      <td>255344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2062324</th>\n",
       "      <td>55350</td>\n",
       "      <td>0</td>\n",
       "      <td>Xenopus tropicalis</td>\n",
       "      <td>255344</td>\n",
       "      <td>0</td>\n",
       "      <td>vus_exact_match_score</td>\n",
       "      <td>255344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2062325</th>\n",
       "      <td>55350</td>\n",
       "      <td>0</td>\n",
       "      <td>Xenopus tropicalis</td>\n",
       "      <td>255344</td>\n",
       "      <td>0</td>\n",
       "      <td>benign_exact_match_score</td>\n",
       "      <td>255344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2062326</th>\n",
       "      <td>55350</td>\n",
       "      <td>552</td>\n",
       "      <td>Danio rerio</td>\n",
       "      <td>255344</td>\n",
       "      <td>167</td>\n",
       "      <td>exact_match_score</td>\n",
       "      <td>255344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2062327</th>\n",
       "      <td>55350</td>\n",
       "      <td>13</td>\n",
       "      <td>Danio rerio</td>\n",
       "      <td>255344</td>\n",
       "      <td>2</td>\n",
       "      <td>ptm_exact_match_score</td>\n",
       "      <td>255344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2062328</th>\n",
       "      <td>55350</td>\n",
       "      <td>335</td>\n",
       "      <td>Danio rerio</td>\n",
       "      <td>255344</td>\n",
       "      <td>101</td>\n",
       "      <td>variant_exact_match_score</td>\n",
       "      <td>255344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2062329</th>\n",
       "      <td>55350</td>\n",
       "      <td>0</td>\n",
       "      <td>Danio rerio</td>\n",
       "      <td>255344</td>\n",
       "      <td>0</td>\n",
       "      <td>pathogenic_exact_match_score</td>\n",
       "      <td>255344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2062330</th>\n",
       "      <td>55350</td>\n",
       "      <td>0</td>\n",
       "      <td>Danio rerio</td>\n",
       "      <td>255344</td>\n",
       "      <td>0</td>\n",
       "      <td>vus_exact_match_score</td>\n",
       "      <td>255344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2062331</th>\n",
       "      <td>55350</td>\n",
       "      <td>0</td>\n",
       "      <td>Danio rerio</td>\n",
       "      <td>255344</td>\n",
       "      <td>0</td>\n",
       "      <td>benign_exact_match_score</td>\n",
       "      <td>255344</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2062332 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         msa_id  number_of_aminoacids pairwise_species_id  \\\n",
       "0            31                   309        Mus musculus   \n",
       "1            31                     0        Mus musculus   \n",
       "2            31                    43        Mus musculus   \n",
       "3            31                     0        Mus musculus   \n",
       "4            31                     0        Mus musculus   \n",
       "5            31                     0        Mus musculus   \n",
       "6            31                   309   Rattus norvegicus   \n",
       "7            31                     0   Rattus norvegicus   \n",
       "8            31                    43   Rattus norvegicus   \n",
       "9            31                     0   Rattus norvegicus   \n",
       "10           31                     0   Rattus norvegicus   \n",
       "11           31                     0   Rattus norvegicus   \n",
       "12           31                   309     Pan troglodytes   \n",
       "13           31                     0     Pan troglodytes   \n",
       "14           31                    43     Pan troglodytes   \n",
       "15           31                     0     Pan troglodytes   \n",
       "16           31                     0     Pan troglodytes   \n",
       "17           31                     0     Pan troglodytes   \n",
       "18           31                   309      Macaca mulatta   \n",
       "19           31                     0      Macaca mulatta   \n",
       "20           31                    43      Macaca mulatta   \n",
       "21           31                     0      Macaca mulatta   \n",
       "22           31                     0      Macaca mulatta   \n",
       "23           31                     0      Macaca mulatta   \n",
       "24           31                   309  Xenopus tropicalis   \n",
       "25           31                     0  Xenopus tropicalis   \n",
       "26           31                    43  Xenopus tropicalis   \n",
       "27           31                     0  Xenopus tropicalis   \n",
       "28           31                     0  Xenopus tropicalis   \n",
       "29           31                     0  Xenopus tropicalis   \n",
       "...         ...                   ...                 ...   \n",
       "2062302   55350                   552     Pan troglodytes   \n",
       "2062303   55350                    13     Pan troglodytes   \n",
       "2062304   55350                   335     Pan troglodytes   \n",
       "2062305   55350                     0     Pan troglodytes   \n",
       "2062306   55350                     0     Pan troglodytes   \n",
       "2062307   55350                     0     Pan troglodytes   \n",
       "2062308   55350                   552      Macaca mulatta   \n",
       "2062309   55350                    13      Macaca mulatta   \n",
       "2062310   55350                   335      Macaca mulatta   \n",
       "2062311   55350                     0      Macaca mulatta   \n",
       "2062312   55350                     0      Macaca mulatta   \n",
       "2062313   55350                     0      Macaca mulatta   \n",
       "2062314   55350                   552        Mus musculus   \n",
       "2062315   55350                    13        Mus musculus   \n",
       "2062316   55350                   335        Mus musculus   \n",
       "2062317   55350                     0        Mus musculus   \n",
       "2062318   55350                     0        Mus musculus   \n",
       "2062319   55350                     0        Mus musculus   \n",
       "2062320   55350                   552  Xenopus tropicalis   \n",
       "2062321   55350                    13  Xenopus tropicalis   \n",
       "2062322   55350                   335  Xenopus tropicalis   \n",
       "2062323   55350                     0  Xenopus tropicalis   \n",
       "2062324   55350                     0  Xenopus tropicalis   \n",
       "2062325   55350                     0  Xenopus tropicalis   \n",
       "2062326   55350                   552         Danio rerio   \n",
       "2062327   55350                    13         Danio rerio   \n",
       "2062328   55350                   335         Danio rerio   \n",
       "2062329   55350                     0         Danio rerio   \n",
       "2062330   55350                     0         Danio rerio   \n",
       "2062331   55350                     0         Danio rerio   \n",
       "\n",
       "        pairwise_transcript_id  score                    score_type  \\\n",
       "0                          371    269             exact_match_score   \n",
       "1                          371      0         ptm_exact_match_score   \n",
       "2                          371     38     variant_exact_match_score   \n",
       "3                          371      0  pathogenic_exact_match_score   \n",
       "4                          371      0         vus_exact_match_score   \n",
       "5                          371      0      benign_exact_match_score   \n",
       "6                          371    269             exact_match_score   \n",
       "7                          371      0         ptm_exact_match_score   \n",
       "8                          371     38     variant_exact_match_score   \n",
       "9                          371      0  pathogenic_exact_match_score   \n",
       "10                         371      0         vus_exact_match_score   \n",
       "11                         371      0      benign_exact_match_score   \n",
       "12                         371    296             exact_match_score   \n",
       "13                         371      0         ptm_exact_match_score   \n",
       "14                         371     42     variant_exact_match_score   \n",
       "15                         371      0  pathogenic_exact_match_score   \n",
       "16                         371      0         vus_exact_match_score   \n",
       "17                         371      0      benign_exact_match_score   \n",
       "18                         371    291             exact_match_score   \n",
       "19                         371      0         ptm_exact_match_score   \n",
       "20                         371     41     variant_exact_match_score   \n",
       "21                         371      0  pathogenic_exact_match_score   \n",
       "22                         371      0         vus_exact_match_score   \n",
       "23                         371      0      benign_exact_match_score   \n",
       "24                         371    228             exact_match_score   \n",
       "25                         371      0         ptm_exact_match_score   \n",
       "26                         371     33     variant_exact_match_score   \n",
       "27                         371      0  pathogenic_exact_match_score   \n",
       "28                         371      0         vus_exact_match_score   \n",
       "29                         371      0      benign_exact_match_score   \n",
       "...                        ...    ...                           ...   \n",
       "2062302                 255344    545             exact_match_score   \n",
       "2062303                 255344     13         ptm_exact_match_score   \n",
       "2062304                 255344    329     variant_exact_match_score   \n",
       "2062305                 255344      0  pathogenic_exact_match_score   \n",
       "2062306                 255344      0         vus_exact_match_score   \n",
       "2062307                 255344      0      benign_exact_match_score   \n",
       "2062308                 255344    522             exact_match_score   \n",
       "2062309                 255344     11         ptm_exact_match_score   \n",
       "2062310                 255344    313     variant_exact_match_score   \n",
       "2062311                 255344      0  pathogenic_exact_match_score   \n",
       "2062312                 255344      0         vus_exact_match_score   \n",
       "2062313                 255344      0      benign_exact_match_score   \n",
       "2062314                 255344    403             exact_match_score   \n",
       "2062315                 255344      9         ptm_exact_match_score   \n",
       "2062316                 255344    233     variant_exact_match_score   \n",
       "2062317                 255344      0  pathogenic_exact_match_score   \n",
       "2062318                 255344      0         vus_exact_match_score   \n",
       "2062319                 255344      0      benign_exact_match_score   \n",
       "2062320                 255344    230             exact_match_score   \n",
       "2062321                 255344      7         ptm_exact_match_score   \n",
       "2062322                 255344    135     variant_exact_match_score   \n",
       "2062323                 255344      0  pathogenic_exact_match_score   \n",
       "2062324                 255344      0         vus_exact_match_score   \n",
       "2062325                 255344      0      benign_exact_match_score   \n",
       "2062326                 255344    167             exact_match_score   \n",
       "2062327                 255344      2         ptm_exact_match_score   \n",
       "2062328                 255344    101     variant_exact_match_score   \n",
       "2062329                 255344      0  pathogenic_exact_match_score   \n",
       "2062330                 255344      0         vus_exact_match_score   \n",
       "2062331                 255344      0      benign_exact_match_score   \n",
       "\n",
       "        transcript_id  \n",
       "0                 371  \n",
       "1                 371  \n",
       "2                 371  \n",
       "3                 371  \n",
       "4                 371  \n",
       "5                 371  \n",
       "6                 371  \n",
       "7                 371  \n",
       "8                 371  \n",
       "9                 371  \n",
       "10                371  \n",
       "11                371  \n",
       "12                371  \n",
       "13                371  \n",
       "14                371  \n",
       "15                371  \n",
       "16                371  \n",
       "17                371  \n",
       "18                371  \n",
       "19                371  \n",
       "20                371  \n",
       "21                371  \n",
       "22                371  \n",
       "23                371  \n",
       "24                371  \n",
       "25                371  \n",
       "26                371  \n",
       "27                371  \n",
       "28                371  \n",
       "29                371  \n",
       "...               ...  \n",
       "2062302        255344  \n",
       "2062303        255344  \n",
       "2062304        255344  \n",
       "2062305        255344  \n",
       "2062306        255344  \n",
       "2062307        255344  \n",
       "2062308        255344  \n",
       "2062309        255344  \n",
       "2062310        255344  \n",
       "2062311        255344  \n",
       "2062312        255344  \n",
       "2062313        255344  \n",
       "2062314        255344  \n",
       "2062315        255344  \n",
       "2062316        255344  \n",
       "2062317        255344  \n",
       "2062318        255344  \n",
       "2062319        255344  \n",
       "2062320        255344  \n",
       "2062321        255344  \n",
       "2062322        255344  \n",
       "2062323        255344  \n",
       "2062324        255344  \n",
       "2062325        255344  \n",
       "2062326        255344  \n",
       "2062327        255344  \n",
       "2062328        255344  \n",
       "2062329        255344  \n",
       "2062330        255344  \n",
       "2062331        255344  \n",
       "\n",
       "[2062332 rows x 7 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.DataFrame(scores)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.DataFrame(scores)\n",
    "\n",
    "def get_id(x):\n",
    "\tglobal ncbi_to_transcript_id\n",
    "\tsearch_params = [x]\n",
    "\n",
    "\tfor param in search_params:\n",
    "\t\tif param in ncbi_to_transcript_id.index:\n",
    "\t\t\treturn ncbi_to_transcript_id.loc[param, 'gene_symbol']\n",
    "\n",
    "\treturn None\n",
    "\n",
    "df.loc[:, 'gene_symbol'] = df['transcript_id'].apply( get_id )\n",
    "\n",
    "column_map = {\n",
    "\t'number_of_aminoacids': 'aminoacid_number',\n",
    "\t'pairwise_species_id': 'specie',\n",
    "\t'pairwise_transcript_id':'transcript_id_specie',\n",
    "\t'score': 'score',\n",
    "\t'score_type': 'score_type',\n",
    "\t'transcript_id': 'transcript_id',\n",
    "\t'gene_symbol': 'gene_symbol',\n",
    "    'msa_id': 'msa_id'\n",
    "}\n",
    "\n",
    "df.columns = [column_map[col] for col in df.columns]\n",
    "df.loc[df['gene_symbol'].isna(), 'gene_symbol'] = ''\n",
    "cur.execute('TRUNCATE TABLE conservation_scores')\n",
    "con.commit()\n",
    "engine = create_engine(\"mysql+pymysql://{user}:{pw}@{server}/{db}\"\n",
    "                       .format(server=config['MYSQL_HOST'], user=config['MYSQL_USER'], pw=config['MYSQL_PASSWD'], \n",
    "                               db=config['MYSQL_DB']))\n",
    "\n",
    "df.to_sql('conservation_scores', engine, if_exists='append', index=False)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
