{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "from sqlalchemy import create_engine\n",
    "import pandas as pd\n",
    "import os\n",
    "from os import path\n",
    "import re\n",
    "import pymysql\n",
    "import matplotlib\n",
    "matplotlib.use('Agg')\n",
    "from matplotlib import pyplot as plt\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "from tqdm import tqdm\n",
    "from multiprocessing import Pool\n",
    "def imap_unordered_bar(func, args, number_of_elements, n_processes = 30):\n",
    "    scores = []\n",
    "    p = Pool(n_processes)\n",
    "\n",
    "    with tqdm(total = number_of_elements) as pbar:\n",
    "        for res in tqdm(p.imap_unordered(func, args, chunksize=n_processes)):\n",
    "            scores += res\n",
    "            pbar.update()\n",
    "\n",
    "    pbar.close()\n",
    "    p.close()\n",
    "    p.join()\n",
    "\n",
    "    return scores\n",
    "\n",
    "engine = create_engine(\"mysql+pymysql://{user}:{pw}@localhost/{db}\"\n",
    "                       .format(user=\"root\",\n",
    "                               pw=\"\",\n",
    "                               db=\"current_project\"))\n",
    "\n",
    "\n",
    "plot_dir = '/opt/current_project/tmp/clinvar/'\n",
    "alignment_dir = \"/opt/current_project/results/alignments/\"\n",
    "\n",
    "polarity = {\n",
    "\t'A': 'nonpolar',\n",
    "    'B': '-',\n",
    "    'Z': '-',\n",
    "\t'R': 'polar',\n",
    "\t'N': 'polar',\n",
    "\t'D': 'polar',\n",
    "\t'C': 'nonpolar',\n",
    "\t'E': 'polar',\n",
    "\t'Q': 'polar',\n",
    "\t'G': 'nonpolar',\n",
    "\t'H': 'polar',\n",
    "\t'I': 'nonpolar',\n",
    "\t'L': 'nonpolar',\n",
    "\t'K': 'polar',\n",
    "\t'M': 'nonpolar',\n",
    "\t'F': 'nonpolar',\n",
    "\t'P': 'nonpolar',\n",
    "\t'S': 'polar',\n",
    "\t'T': 'polar',\n",
    "\t'W': 'nonpolar',\n",
    "\t'Y': 'polar',\n",
    "\t'V': 'nonpolar',\n",
    "\t'U': 'polar',\n",
    "\t'-': '-'\n",
    "}\n",
    "def sqlize(significance_list):\n",
    "\treturn \"'\"+ \"', '\".join(significance_list) + \"'\"\n",
    "\n",
    "con = pymysql.connect(host='127.0.0.1', unix_socket='/opt/lampp/var/mysql/mysql.sock', user='root', passwd='', db='current_project')\n",
    "cur = con.cursor()\n",
    "\n",
    "ncbi_to_transcript_id = pd.read_sql(\"SELECT gdb.db_id AS NCBI, gdb.convart_gene_id AS id, ncbi.meta_value AS gene_symbol \"\n",
    "\t\t\t\t\t\t\t\t\"FROM convart_gene_to_db AS gdb \" +\n",
    "\t\t\t\t\t\t\t\t\"LEFT JOIN ncbi_gene_meta AS ncbi ON ncbi.ncbi_gene_id = gdb.db_id WHERE gdb.db='NCBI' \"+\n",
    "\t\t\t\t\t\t\t\t\"AND ncbi.meta_key = 'gene_symbol' GROUP BY gdb.convart_gene_id\", con)\n",
    "ncbi_to_transcript_id.set_index('id', drop=False, inplace=True)\n",
    "\n",
    "\n",
    "genes = pd.read_sql(\"SELECT DISTINCT g.id, fasta, sequence, m.id as msa_id \"+\n",
    "\t\t\t\"FROM convart_gene AS g \"+\n",
    "\t\t\t\"INNER JOIN msa_best_combination AS mb ON mb.convart_gene_id = g.id \" +\n",
    "\t\t\t\"INNER JOIN msa AS m ON m.id = mb.msa_id \"\n",
    "\t\t\t\"WHERE g.species_id='Homo sapiens'\", con)\n",
    "\n",
    "\n",
    "benign_list = sqlize([\"Benign\",\"Benign, association\",\"Benign/Likely benign\",\"Benign/Likely benign, Affects\",\"Benign/Likely benign, association\",\n",
    "\t\t\t\"Benign/Likely benign, drug response\",\"Benign/Likely benign, drug response, risk factor\",\"Benign/Likely benign, other\",\"Benign/Likely benign, protective\",\n",
    "\t\t\t\"Benign/Likely benign, protective, risk factor\",\"Benign/Likely benign, risk factor\",\"Benign, other\",\"Benign, risk factor\",\"Likely benign\",\n",
    "\t\t\t\"Likely benign, drug response, other\",\"Likely benign, other\",\"Likely benign, risk factor\"])\n",
    "vus_list = sqlize([\"Uncertain significance\",\"Uncertain significance, drug response\",\"Uncertain significance, other\",\"Uncertain significance, risk factor\"])\n",
    "pathogenic_list = sqlize([\"Pathogenic\",\"Pathogenic, Affects\",\"Pathogenic, association, protective\",\"Pathogenic, drug response\",\"Pathogenic/Likely pathogenic\",\n",
    "\"Pathogenic/Likely pathogenic, drug response\",\"Pathogenic/Likely pathogenic, other\",\"Pathogenic/Likely pathogenic, risk factor\",\"Pathogenic, other\",\n",
    "\"Pathogenic, other, risk factor\",\"Pathogenic, protective\",\"Pathogenic, risk factor\",\"Likely pathogenic\",\"Likely pathogenic, association\",\n",
    "\"Likely pathogenic, drug response\",\"Likely pathogenic, other\",\"Likely pathogenic, risk factor\"])\n",
    "\n",
    "ptm_data = pd.read_sql(\"SELECT cdb.convart_gene_id AS transcript_id, GROUP_CONCAT(DISTINCT position) as positions\"\n",
    "\t\t\t\t\t+\", COUNT(DISTINCT position) as variation_count FROM ptm INNER JOIN convart_gene_to_db AS cdb ON cdb.db_id=ptm.acc_id  WHERE position > 0 \"\n",
    "                    +\"GROUP BY cdb.convart_gene_id\", con)\n",
    "clinvar_data = pd.read_sql(\"SELECT gdb.convart_gene_id AS transcript_id, GROUP_CONCAT(DISTINCT position) as positions\"\n",
    "\t\t\t+\", COUNT(DISTINCT position) as variation_count FROM clinvar INNER JOIN convart_gene_to_db AS gdb ON gdb.db_id=np_id WHERE position > 0 \"\n",
    "                           +\"GROUP BY gdb.convart_gene_id\", con)\n",
    "pathogenic_clinvar_data = pd.read_sql(\"SELECT gdb.convart_gene_id AS transcript_id, GROUP_CONCAT(DISTINCT position) as positions, COUNT(DISTINCT position) \"\n",
    "\t\t\t+\"as variation_count FROM clinvar INNER JOIN convart_gene_to_db AS gdb ON gdb.db_id=np_id WHERE position > 0 AND clinical_significance IN (\"+pathogenic_list+\")\"+\n",
    "\t\t\t\" GROUP BY gdb.convart_gene_id\", con)\n",
    "vus_clinvar_data = pd.read_sql(\"SELECT gdb.convart_gene_id AS transcript_id, GROUP_CONCAT(DISTINCT position) as positions, COUNT(DISTINCT position) \"\n",
    "\t\t\t+\"as variation_count FROM clinvar INNER JOIN convart_gene_to_db AS gdb ON gdb.db_id=np_id WHERE position > 0 AND clinical_significance IN (\"+vus_list+\")\"+\n",
    "\t\t\t\" GROUP BY gdb.convart_gene_id\", con)\n",
    "benign_clinvar_data = pd.read_sql(\"SELECT gdb.convart_gene_id AS transcript_id, GROUP_CONCAT(DISTINCT position) as positions, COUNT(DISTINCT position) \"\n",
    "\t\t\t+\"as variation_count FROM clinvar INNER JOIN convart_gene_to_db AS gdb ON gdb.db_id=np_id WHERE position > 0 AND clinical_significance IN (\"+benign_list+\")\"+\n",
    "\t\t\t\" GROUP BY gdb.convart_gene_id\", con)\n",
    "gnomad_data = pd.read_sql(\"SELECT gdb.convart_gene_id AS transcript_id, GROUP_CONCAT(DISTINCT position) as positions\"\n",
    "\t\t\t\t\t+\", COUNT(DISTINCT position) as variation_count FROM gnomad INNER JOIN convart_gene_to_db AS gdb ON gdb.db_id=canonical_transcript WHERE position > 0 \"+\n",
    "                     \" GROUP BY gdb.convart_gene_id\", con)\n",
    "cosmic_data = pd.read_sql(\"SELECT gdb.convart_gene_id AS transcript_id, GROUP_CONCAT(DISTINCT position) as positions\"\n",
    "\t\t\t\t\t+\", COUNT(DISTINCT position) as variation_count FROM CosmicMutantExport INNER JOIN convart_gene_to_db AS gdb ON gdb.db_id=accession_number WHERE position > 0 \"+\n",
    "                     \" GROUP BY gdb.convart_gene_id\", con)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_data = pd.concat([clinvar_data, gnomad_data, cosmic_data], axis=0, ignore_index=True)\n",
    "combined_data = combined_data.groupby('transcript_id').apply(lambda x: pd.Series({'positions':','.join([str(i) for i in x['positions']]), \n",
    "                                                                  'variation_count': sum(x['variation_count'])}))\n",
    "combined_data = combined_data.reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|          | 0/55428 [00:00<?, ?it/s]\n",
      "  0%|          | 1/55428 [00:00<7:45:58,  1.98it/s]\n",
      "  0%|          | 121/55428 [00:00<5:25:44,  2.83it/s]\n",
      "  1%|          | 451/55428 [00:00<3:46:44,  4.04it/s]\n",
      "  1%|          | 691/55428 [00:00<2:38:08,  5.77it/s]\n",
      "  2%|▏         | 961/55428 [00:00<1:50:15,  8.23it/s]\n",
      "  2%|▏         | 1171/55428 [00:01<1:17:02, 11.74it/s]\n",
      "  3%|▎         | 1501/55428 [00:01<53:41, 16.74it/s]  \n",
      "  3%|▎         | 1718/55428 [00:01<37:33, 23.83it/s]\n",
      "  3%|▎         | 1934/55428 [00:01<26:20, 33.85it/s]\n",
      "  4%|▍         | 2221/55428 [00:01<18:26, 48.09it/s]\n",
      "  5%|▍         | 2551/55428 [00:01<12:54, 68.23it/s]\n",
      "  5%|▌         | 2796/55428 [00:01<09:08, 96.02it/s]\n",
      "  5%|▌         | 3025/55428 [00:01<06:29, 134.55it/s]\n",
      "  6%|▌         | 3249/55428 [00:01<04:42, 184.70it/s]\n",
      "  6%|▋         | 3482/55428 [00:02<03:24, 253.66it/s]\n",
      "  7%|▋         | 3811/55428 [00:02<02:27, 349.80it/s]\n",
      "  7%|▋         | 4081/55428 [00:02<01:49, 468.74it/s]\n",
      "  8%|▊         | 4310/55428 [00:02<01:24, 607.62it/s]\n",
      "  8%|▊         | 4561/55428 [00:02<01:05, 771.79it/s]\n",
      "  9%|▊         | 4831/55428 [00:02<00:51, 977.42it/s]\n",
      "  9%|▉         | 5161/55428 [00:02<00:41, 1220.08it/s]\n",
      " 10%|▉         | 5409/55428 [00:02<00:36, 1367.42it/s]\n",
      " 10%|█         | 5640/55428 [00:03<00:32, 1512.42it/s]\n",
      " 11%|█         | 5862/55428 [00:03<00:33, 1465.51it/s]\n",
      " 11%|█         | 6151/55428 [00:03<00:28, 1716.89it/s]\n",
      " 12%|█▏        | 6391/55428 [00:03<00:27, 1809.84it/s]\n",
      " 12%|█▏        | 6610/55428 [00:03<00:26, 1811.71it/s]\n",
      " 12%|█▏        | 6818/55428 [00:03<00:30, 1568.73it/s]\n",
      " 13%|█▎        | 7201/55428 [00:03<00:26, 1849.03it/s]\n",
      " 13%|█▎        | 7441/55428 [00:03<00:25, 1885.98it/s]\n",
      " 14%|█▍        | 7656/55428 [00:04<00:27, 1736.22it/s]\n",
      " 14%|█▍        | 7851/55428 [00:04<00:29, 1600.40it/s]\n",
      " 15%|█▍        | 8251/55428 [00:04<00:24, 1918.62it/s]\n",
      " 15%|█▌        | 8482/55428 [00:04<00:24, 1931.13it/s]\n",
      " 16%|█▌        | 8703/55428 [00:04<00:23, 1969.47it/s]\n",
      " 16%|█▌        | 8920/55428 [00:04<00:24, 1933.93it/s]\n",
      " 16%|█▋        | 9128/55428 [00:04<00:25, 1823.86it/s]\n",
      " 17%|█▋        | 9601/55428 [00:05<00:23, 1935.60it/s]\n",
      " 18%|█▊        | 9871/55428 [00:05<00:21, 2089.46it/s]\n",
      " 18%|█▊        | 10171/55428 [00:05<00:20, 2242.09it/s]\n",
      " 19%|█▉        | 10441/55428 [00:05<00:19, 2310.64it/s]\n",
      "10441it [00:05, 2226.80it/s]\u001b[A\n",
      " 19%|█▉        | 10679/55428 [00:05<00:22, 1961.35it/s]\n",
      " 20%|██        | 11276/55428 [00:05<00:20, 2139.55it/s]\n",
      " 21%|██        | 11502/55428 [00:05<00:22, 1924.66it/s]\n",
      " 21%|██▏       | 11821/55428 [00:06<00:19, 2181.74it/s]\n",
      " 22%|██▏       | 12091/55428 [00:06<00:19, 2175.89it/s]\n",
      " 22%|██▏       | 12331/55428 [00:06<00:19, 2228.80it/s]\n",
      " 23%|██▎       | 12565/55428 [00:06<00:20, 2093.87it/s]\n",
      " 23%|██▎       | 12784/55428 [00:06<00:21, 2009.81it/s]\n",
      " 23%|██▎       | 13021/55428 [00:06<00:20, 2030.73it/s]\n",
      " 24%|██▍       | 13351/55428 [00:06<00:18, 2263.04it/s]\n",
      "13351it [00:06, 2269.38it/s]\u001b[A\n",
      " 25%|██▍       | 13590/55428 [00:06<00:22, 1871.69it/s]\n",
      " 25%|██▌       | 13891/55428 [00:07<00:20, 2063.48it/s]\n",
      " 26%|██▌       | 14191/55428 [00:07<00:19, 2140.00it/s]\n",
      " 26%|██▌       | 14431/55428 [00:07<00:18, 2186.83it/s]\n",
      " 26%|██▋       | 14661/55428 [00:07<00:21, 1853.68it/s]\n",
      " 27%|██▋       | 14971/55428 [00:07<00:20, 2018.86it/s]\n",
      " 27%|██▋       | 15189/55428 [00:07<00:19, 2026.21it/s]\n",
      " 28%|██▊       | 15481/55428 [00:07<00:19, 2069.05it/s]\n",
      " 28%|██▊       | 15721/55428 [00:07<00:18, 2121.26it/s]\n",
      " 29%|██▉       | 15991/55428 [00:08<00:18, 2110.67it/s]\n",
      "15991it [00:07, 2111.36it/s]\u001b[A\n",
      " 29%|██▉       | 16207/55428 [00:08<00:21, 1814.20it/s]\n",
      " 30%|███       | 16741/55428 [00:08<00:17, 2260.38it/s]\n",
      " 31%|███       | 17035/55428 [00:08<00:19, 1926.43it/s]\n",
      "17060it [00:08, 1972.12it/s]\u001b[A\n",
      " 32%|███▏      | 17581/55428 [00:08<00:19, 1928.82it/s]\n",
      "17581it [00:08, 1950.73it/s]\u001b[A\n",
      " 32%|███▏      | 17804/55428 [00:08<00:19, 1884.23it/s]\n",
      " 33%|███▎      | 18062/55428 [00:09<00:20, 1825.05it/s]\n",
      " 33%|███▎      | 18421/55428 [00:09<00:17, 2124.63it/s]\n",
      " 34%|███▎      | 18663/55428 [00:09<00:19, 1919.89it/s]\n",
      " 34%|███▍      | 18901/55428 [00:09<00:18, 2025.19it/s]\n",
      " 35%|███▍      | 19141/55428 [00:09<00:17, 2086.98it/s]\n",
      " 35%|███▍      | 19364/55428 [00:09<00:18, 1924.71it/s]\n",
      " 35%|███▌      | 19569/55428 [00:09<00:18, 1924.37it/s]\n",
      " 36%|███▌      | 19831/55428 [00:09<00:17, 2051.26it/s]\n",
      " 36%|███▌      | 20045/55428 [00:09<00:17, 2058.37it/s]\n",
      " 37%|███▋      | 20281/55428 [00:10<00:16, 2115.14it/s]\n",
      " 37%|███▋      | 20497/55428 [00:10<00:18, 1899.30it/s]\n",
      " 38%|███▊      | 20792/55428 [00:10<00:16, 2046.28it/s]\n",
      " 38%|███▊      | 21151/55428 [00:10<00:14, 2316.50it/s]\n",
      " 39%|███▊      | 21401/55428 [00:10<00:18, 1875.27it/s]\n",
      " 39%|███▉      | 21631/55428 [00:10<00:17, 1900.27it/s]\n",
      " 39%|███▉      | 21871/55428 [00:10<00:17, 1939.63it/s]\n",
      " 40%|███▉      | 22111/55428 [00:11<00:18, 1803.53it/s]\n",
      " 40%|████      | 22381/55428 [00:11<00:16, 1975.59it/s]\n",
      " 41%|████      | 22681/55428 [00:11<00:15, 2166.22it/s]\n",
      " 41%|████▏     | 22913/55428 [00:11<00:16, 1964.66it/s]\n",
      " 42%|████▏     | 23124/55428 [00:11<00:17, 1833.73it/s]\n",
      " 42%|████▏     | 23319/55428 [00:11<00:17, 1848.45it/s]\n",
      " 42%|████▏     | 23522/55428 [00:11<00:17, 1854.75it/s]\n",
      " 43%|████▎     | 23821/55428 [00:11<00:16, 1964.50it/s]\n",
      " 43%|████▎     | 24024/55428 [00:11<00:15, 1971.53it/s]\n",
      " 44%|████▎     | 24226/55428 [00:12<00:16, 1933.52it/s]\n",
      " 45%|████▍     | 24781/55428 [00:12<00:14, 2144.46it/s]\n",
      " 45%|████▌     | 25081/55428 [00:12<00:14, 2104.62it/s]\n",
      "25081it [00:12, 2006.35it/s]\u001b[A\n",
      " 46%|████▌     | 25305/55428 [00:12<00:18, 1649.09it/s]\n",
      " 46%|████▋     | 25771/55428 [00:12<00:14, 1984.99it/s]\n",
      " 47%|████▋     | 26071/55428 [00:12<00:13, 2130.57it/s]\n",
      " 47%|████▋     | 26321/55428 [00:12<00:13, 2186.61it/s]\n",
      " 48%|████▊     | 26566/55428 [00:13<00:15, 1856.36it/s]\n",
      "26570it [00:12, 1902.38it/s]\u001b[A\n",
      " 48%|████▊     | 26779/55428 [00:13<00:15, 1814.15it/s]\n",
      " 49%|████▉     | 27301/55428 [00:13<00:13, 2109.26it/s]\n",
      " 50%|████▉     | 27601/55428 [00:13<00:12, 2190.41it/s]\n",
      " 50%|█████     | 27833/55428 [00:13<00:13, 2059.95it/s]\n",
      " 51%|█████     | 28081/55428 [00:13<00:12, 2126.81it/s]\n",
      " 51%|█████     | 28302/55428 [00:13<00:13, 2077.07it/s]\n",
      " 52%|█████▏    | 28561/55428 [00:14<00:12, 2151.18it/s]\n",
      " 52%|█████▏    | 28781/55428 [00:14<00:14, 1867.25it/s]\n",
      " 52%|█████▏    | 29011/55428 [00:14<00:13, 1956.02it/s]\n",
      " 53%|█████▎    | 29341/55428 [00:14<00:12, 2109.13it/s]\n",
      " 53%|█████▎    | 29611/55428 [00:14<00:11, 2240.03it/s]\n",
      " 54%|█████▍    | 29851/55428 [00:14<00:11, 2193.97it/s]\n",
      " 54%|█████▍    | 30121/55428 [00:14<00:11, 2190.75it/s]\n",
      " 55%|█████▍    | 30345/55428 [00:14<00:11, 2147.98it/s]\n",
      " 55%|█████▌    | 30564/55428 [00:15<00:12, 1915.22it/s]\n",
      " 56%|█████▌    | 30821/55428 [00:15<00:11, 2073.72it/s]\n",
      " 56%|█████▌    | 31038/55428 [00:15<00:12, 1882.37it/s]\n",
      " 57%|█████▋    | 31351/55428 [00:15<00:11, 2104.96it/s]\n",
      " 57%|█████▋    | 31621/55428 [00:15<00:10, 2236.20it/s]\n",
      " 57%|█████▋    | 31858/55428 [00:15<00:11, 2135.32it/s]\n",
      "31858it [00:15, 2096.69it/s]\u001b[A\n",
      " 58%|█████▊    | 32082/55428 [00:15<00:12, 1888.16it/s]\n",
      " 58%|█████▊    | 32341/55428 [00:15<00:11, 2028.90it/s]\n",
      " 59%|█████▊    | 32556/55428 [00:16<00:11, 1993.39it/s]\n",
      " 59%|█████▉    | 32765/55428 [00:16<00:12, 1834.17it/s]\n",
      " 60%|█████▉    | 33061/55428 [00:16<00:10, 2056.68it/s]\n",
      " 60%|██████    | 33282/55428 [00:16<00:10, 2035.00it/s]\n",
      " 60%|██████    | 33496/55428 [00:16<00:10, 2032.42it/s]\n",
      " 61%|██████    | 33707/55428 [00:16<00:11, 1893.00it/s]\n",
      " 61%|██████    | 33904/55428 [00:16<00:12, 1786.77it/s]\n",
      " 62%|██████▏   | 34261/55428 [00:16<00:10, 2061.19it/s]\n",
      " 62%|██████▏   | 34487/55428 [00:16<00:10, 1932.66it/s]\n",
      " 63%|██████▎   | 34771/55428 [00:17<00:09, 2128.68it/s]\n",
      " 63%|██████▎   | 35001/55428 [00:17<00:09, 2152.35it/s]\n",
      " 64%|██████▎   | 35229/55428 [00:17<00:09, 2079.42it/s]\n",
      " 64%|██████▍   | 35446/55428 [00:17<00:09, 2085.83it/s]\n",
      " 64%|██████▍   | 35672/55428 [00:17<00:09, 2132.49it/s]\n",
      " 65%|██████▌   | 36133/55428 [00:17<00:09, 2055.19it/s]\n",
      " 66%|██████▌   | 36481/55428 [00:17<00:08, 2234.38it/s]\n",
      "36481it [00:17, 2063.84it/s]\u001b[A\n",
      " 66%|██████▌   | 36712/55428 [00:18<00:10, 1822.62it/s]\n",
      " 67%|██████▋   | 36930/55428 [00:18<00:09, 1908.72it/s]\n",
      " 67%|██████▋   | 37202/55428 [00:18<00:08, 2073.16it/s]\n",
      " 68%|██████▊   | 37423/55428 [00:18<00:09, 1878.71it/s]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 68%|██████▊   | 37771/55428 [00:18<00:08, 2144.69it/s]\n",
      " 69%|██████▊   | 38008/55428 [00:18<00:08, 2079.09it/s]\n",
      " 69%|██████▉   | 38232/55428 [00:18<00:08, 2069.04it/s]\n",
      " 69%|██████▉   | 38451/55428 [00:18<00:09, 1861.38it/s]\n",
      " 70%|███████   | 38881/55428 [00:18<00:07, 2219.65it/s]\n",
      " 71%|███████   | 39141/55428 [00:19<00:08, 1982.11it/s]\n",
      " 71%|███████   | 39371/55428 [00:19<00:08, 1922.79it/s]\n",
      " 71%|███████▏  | 39586/55428 [00:19<00:08, 1779.79it/s]\n",
      " 72%|███████▏  | 39901/55428 [00:19<00:07, 2026.70it/s]\n",
      " 72%|███████▏  | 40128/55428 [00:19<00:07, 2033.76it/s]\n",
      " 73%|███████▎  | 40349/55428 [00:19<00:07, 2064.88it/s]\n",
      " 73%|███████▎  | 40568/55428 [00:19<00:07, 1858.55it/s]\n",
      " 74%|███████▎  | 40767/55428 [00:19<00:07, 1889.97it/s]\n",
      " 74%|███████▍  | 40981/55428 [00:20<00:07, 1842.31it/s]\n",
      " 74%|███████▍  | 41221/55428 [00:20<00:07, 1964.42it/s]\n",
      " 75%|███████▍  | 41425/55428 [00:20<00:07, 1928.10it/s]\n",
      " 75%|███████▌  | 41623/55428 [00:20<00:07, 1911.64it/s]\n",
      " 76%|███████▌  | 41941/55428 [00:20<00:06, 2150.73it/s]\n",
      " 76%|███████▌  | 42169/55428 [00:20<00:06, 2100.89it/s]\n",
      " 76%|███████▋  | 42388/55428 [00:20<00:06, 2120.15it/s]\n",
      " 77%|███████▋  | 42631/55428 [00:20<00:06, 2092.86it/s]\n",
      " 77%|███████▋  | 42845/55428 [00:20<00:06, 2081.65it/s]\n",
      " 78%|███████▊  | 43057/55428 [00:21<00:06, 2015.35it/s]\n",
      " 78%|███████▊  | 43351/55428 [00:21<00:05, 2110.65it/s]\n",
      " 79%|███████▊  | 43566/55428 [00:21<00:05, 2025.56it/s]\n",
      " 79%|███████▉  | 43772/55428 [00:21<00:06, 1788.36it/s]\n",
      " 80%|███████▉  | 44101/55428 [00:21<00:05, 2031.31it/s]\n",
      " 80%|███████▉  | 44341/55428 [00:21<00:05, 1909.20it/s]\n",
      " 80%|████████  | 44545/55428 [00:21<00:06, 1710.86it/s]\n",
      " 81%|████████▏ | 45061/55428 [00:21<00:04, 2076.94it/s]\n",
      "45061it [00:21, 2103.58it/s]\u001b[A\n",
      " 82%|████████▏ | 45315/55428 [00:22<00:05, 2001.28it/s]\n",
      " 82%|████████▏ | 45548/55428 [00:22<00:05, 1866.25it/s]\n",
      " 83%|████████▎ | 45760/55428 [00:22<00:05, 1802.05it/s]\n",
      " 83%|████████▎ | 46082/55428 [00:22<00:04, 2030.99it/s]\n",
      " 84%|████████▎ | 46321/55428 [00:22<00:04, 2103.38it/s]\n",
      " 84%|████████▍ | 46555/55428 [00:22<00:04, 1940.83it/s]\n",
      " 84%|████████▍ | 46802/55428 [00:22<00:04, 1965.24it/s]\n",
      " 85%|████████▌ | 47131/55428 [00:22<00:03, 2167.13it/s]\n",
      " 86%|████████▌ | 47401/55428 [00:23<00:03, 2196.76it/s]\n",
      " 86%|████████▌ | 47701/55428 [00:23<00:03, 2315.92it/s]\n",
      " 86%|████████▋ | 47941/55428 [00:23<00:03, 2025.44it/s]\n",
      " 87%|████████▋ | 48211/55428 [00:23<00:03, 2177.19it/s]\n",
      " 87%|████████▋ | 48481/55428 [00:23<00:03, 1969.08it/s]\n",
      " 88%|████████▊ | 48841/55428 [00:23<00:02, 2254.41it/s]\n",
      "48841it [00:23, 2121.80it/s]\u001b[A\n",
      " 89%|████████▊ | 49090/55428 [00:23<00:03, 1864.09it/s]\n",
      " 89%|████████▉ | 49304/55428 [00:24<00:03, 1922.72it/s]\n",
      " 89%|████████▉ | 49561/55428 [00:24<00:02, 2034.12it/s]\n",
      " 90%|█████████ | 50101/55428 [00:24<00:02, 2179.62it/s]\n",
      "50101it [00:24, 2112.18it/s]\u001b[A\n",
      " 91%|█████████ | 50336/55428 [00:24<00:02, 1964.13it/s]\n",
      " 91%|█████████▏| 50671/55428 [00:24<00:02, 2190.71it/s]\n",
      " 92%|█████████▏| 50971/55428 [00:24<00:01, 2329.80it/s]\n",
      " 92%|█████████▏| 51221/55428 [00:24<00:02, 2001.33it/s]\n",
      " 93%|█████████▎| 51441/55428 [00:25<00:01, 2039.18it/s]\n",
      " 93%|█████████▎| 51722/55428 [00:25<00:01, 2039.21it/s]\n",
      " 94%|█████████▍| 52173/55428 [00:25<00:01, 2078.64it/s]\n",
      " 95%|█████████▍| 52386/55428 [00:25<00:01, 1784.80it/s]\n",
      " 95%|█████████▌| 52771/55428 [00:25<00:01, 2124.47it/s]\n",
      " 96%|█████████▌| 53019/55428 [00:25<00:01, 2212.53it/s]\n",
      " 96%|█████████▌| 53266/55428 [00:25<00:00, 2186.51it/s]\n",
      " 97%|█████████▋| 53503/55428 [00:25<00:00, 2025.87it/s]\n",
      " 97%|█████████▋| 53721/55428 [00:26<00:00, 1898.58it/s]\n",
      " 97%|█████████▋| 53941/55428 [00:26<00:00, 1853.06it/s]\n",
      " 98%|█████████▊| 54135/55428 [00:26<00:00, 1785.63it/s]\n",
      " 98%|█████████▊| 54541/55428 [00:26<00:00, 2074.11it/s]\n",
      "54541it [00:26, 2118.50it/s]\u001b[A\n",
      " 99%|█████████▉| 54771/55428 [00:26<00:00, 1733.19it/s]\n",
      "100%|█████████▉| 55297/55428 [00:27<00:00, 1160.29it/s]\n",
      "55298it [00:26, 1177.08it/s]\u001b[A\n",
      "100%|██████████| 55428/55428 [00:27<00:00, 2018.30it/s]\n"
     ]
    }
   ],
   "source": [
    "\n",
    "variations = {\n",
    "    'ptm': ptm_data,\n",
    "\t'variant': combined_data,\n",
    "\t'pathogenic': pathogenic_clinvar_data,\n",
    "\t'vus': vus_clinvar_data,\n",
    "\t'benign': benign_clinvar_data\n",
    "}\n",
    "\n",
    "for key, variation in variations.items():\n",
    "\tvariation.loc[:, 'transcript_id'] = variation.loc[:, 'transcript_id'].astype(str) \n",
    "\tvariations[key] = variation.set_index('transcript_id', drop=False)\n",
    "\n",
    "def conservation_scores_by_gene(row):\n",
    "\tscores = []\n",
    "\n",
    "\tind, msa_details = row\n",
    "   \n",
    "\tmsa_details = msa_details.to_dict()\n",
    "\tmsa_details['id'] = str(msa_details['id'])\n",
    "    \n",
    "\tsequences = ('\\n' + msa_details['fasta']).split('\\n>')\n",
    "\tsequences = [{\n",
    "\t\t\t'transcript_id': msa_details['id'],\n",
    "\t\t\t'species_id': re.sub(r'^(.*?)\\[([a-zA-Z0-9 ]+?)\\]$', r'\\2', seq.split('\\n')[0]),\n",
    "\t\t\t'fasta': re.sub('^(.*?)\\n', '', seq).replace('\\n','').replace(' ', '')} for seq in sequences]\n",
    "\tsequences = list(filter(lambda x : x['fasta'].replace('\\n','') !='', sequences))\n",
    "\n",
    "\thuman_seq = sequences[0]['fasta']\n",
    "\tseq_length = len(human_seq)\n",
    "\tn_sequences = len(sequences)\n",
    "\t\n",
    "\tvariation_positions_by_data = {}\n",
    "\n",
    "\tfor key, data in variations.items():\n",
    "\t\tvariation_positions_by_data[key] = set()\n",
    "        \n",
    "\t\tif str(msa_details['id']) in data.index:\n",
    "\t\t\tcur_positions = data.loc[msa_details['id'], 'positions'].split(',')\n",
    "\t\t\tcur_positions = set([cur_position for cur_position in cur_positions ])\n",
    "\t\t\t[variation_positions_by_data[key].add(int(cur_position))  for cur_position in cur_positions if cur_position != '']\n",
    "        \n",
    "    \n",
    "\tfor pairwise_id in range(1, len(sequences)):\n",
    "\t\texact_match_score = 0\n",
    "\t\tpolarity_match_score = 0\n",
    "\t\t\n",
    "\t\texact_scores_by_data = {}\n",
    "\t\tpolarity_scores_by_data = {}\n",
    "\n",
    "\t\tfor key, data in variations.items():\n",
    "\t\t\texact_scores_by_data[key] = 0\n",
    "\t\t\tpolarity_scores_by_data[key] = 0\n",
    "\n",
    "\t\tpair = sequences[pairwise_id]\n",
    "\t\tpair_seq = pair['fasta']\n",
    "\t\thuman_seq_position = 0\n",
    "\t\t\n",
    "\t\tfor i in range(seq_length):\n",
    "\t\t\tif len(human_seq) != len(pair_seq):\n",
    "\t\t\t\tprint(len(human_seq), len(pair_seq), msa_details['msa_id'])\n",
    "\t\t\t\tsys.exit(-1)\n",
    "\t\t\tif human_seq[i] == '-':\n",
    "\t\t\t\tcontinue\n",
    "\t\t\thuman_seq_position += 1\n",
    "\n",
    "\t\t\tif human_seq[i] == 'X' or pair_seq[i] == 'X':\n",
    "\t\t\t\tcontinue\n",
    "\n",
    "\t\t\taligned_aminoacids = 1 if len({seq[i] for seq in [human_seq, pair_seq]})==1 else 0\n",
    "\t\t\texact_match_score += aligned_aminoacids\n",
    "\n",
    "\t\t\taligned_aminoacids_polarity = 1 if  len({polarity[seq[i]] for seq in [human_seq, pair_seq]}) == 1 else 0\n",
    "\t\t\tpolarity_match_score += aligned_aminoacids_polarity\n",
    "\n",
    "\t\t\tfor key, positions in variation_positions_by_data.items():\n",
    "\t\t\t\tif human_seq_position not in positions:\n",
    "\t\t\t\t\tcontinue\n",
    "\t\t\t\texact_scores_by_data[key] += aligned_aminoacids\n",
    "\t\t\t\tpolarity_scores_by_data[key] += aligned_aminoacids_polarity\n",
    "\n",
    "\t\tscores.append({'transcript_id': msa_details['id'], 'pairwise_transcript_id': pair['transcript_id'],\n",
    "\t\t\t\t\t\t'pairwise_species_id': pair['species_id'],\n",
    "\t\t\t\t\t\t'score_type': 'exact_match_score', 'score': exact_match_score, \n",
    "\t\t\t\t\t\t'number_of_aminoacids': len(msa_details['sequence']),\n",
    "                      'msa_id': msa_details['msa_id']})\n",
    "\t\t#scores.append({'transcript_id': msa_details['id'], 'pairwise_transcript_id': pair['transcript_id'],\n",
    "\t\t#\t\t\t\t'pairwise_species_id': pair['species_id'],\n",
    "\t\t#\t\t\t\t'score_type': 'polarity_match_score', 'score': polarity_match_score,\n",
    "\t\t#\t\t\t\t'number_of_aminoacids': len(msa_details['sequence'])})\n",
    "\n",
    "\t\t#Variant scores\n",
    "\t\tfor key, exact_score in exact_scores_by_data.items():\n",
    "\t\t\tscores.append({'transcript_id': msa_details['id'], 'pairwise_transcript_id': pair['transcript_id'],\n",
    "\t\t\t\t\t\t'pairwise_species_id': pair['species_id'],\n",
    "\t\t\t\t\t\t'score_type': key+'_exact_match_score', 'score': exact_score, \n",
    "\t\t\t\t\t\t'number_of_aminoacids': len(variation_positions_by_data[key]),\n",
    "                      'msa_id': msa_details['msa_id']})\n",
    "\t\t#for key, polarity_score in polarity_scores_by_data.items():\n",
    "\t\t#\tscores.append({'transcript_id': msa_details['id'], 'pairwise_transcript_id': pair['transcript_id'],\n",
    "\t\t#\t\t\t\t'pairwise_species_id': pair['species_id'],\n",
    "\t\t#\t\t\t\t'score_type': key+'_polarity_match_score', 'score': polarity_score, \n",
    "\t\t#\t\t\t\t'number_of_aminoacids': len(variation_positions_by_data[key])})\n",
    "\n",
    "\treturn scores\n",
    "\n",
    "#conservation_scores_by_gene(list(genes.iterrows())[0])\n",
    "scores = imap_unordered_bar(conservation_scores_by_gene, genes.iterrows(), genes.shape[0], 30)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "df = pd.DataFrame(scores)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.DataFrame(scores)\n",
    "\n",
    "def get_id(x):\n",
    "\tglobal ncbi_to_transcript_id\n",
    "\tsearch_params = [x]\n",
    "\n",
    "\tfor param in search_params:\n",
    "\t\tif param in ncbi_to_transcript_id.index:\n",
    "\t\t\treturn ncbi_to_transcript_id.loc[param, 'gene_symbol']\n",
    "\n",
    "\treturn None\n",
    "\n",
    "df.loc[:, 'gene_symbol'] = df['transcript_id'].apply( get_id )\n",
    "\n",
    "column_map = {\n",
    "\t'number_of_aminoacids': 'aminoacid_number',\n",
    "\t'pairwise_species_id': 'specie',\n",
    "\t'pairwise_transcript_id':'transcript_id_specie',\n",
    "\t'score': 'score',\n",
    "\t'score_type': 'score_type',\n",
    "\t'transcript_id': 'transcript_id',\n",
    "\t'gene_symbol': 'gene_symbol',\n",
    "    'msa_id': 'msa_id'\n",
    "}\n",
    "\n",
    "df.columns = [column_map[col] for col in df.columns]\n",
    "df.loc[df['gene_symbol'].isna(), 'gene_symbol'] = ''\n",
    "cur.execute('TRUNCATE TABLE conservation_scores')\n",
    "con.commit()\n",
    "engine = create_engine(\"mysql+pymysql://{user}:{pw}@localhost/{db}\"\n",
    "                       .format(user=\"root\",\n",
    "                               pw=\"\",\n",
    "                               db=\"current_project\"))\n",
    "\n",
    "df.to_sql('conservation_scores', engine, if_exists='append', index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
